---
title: 'Default, simple and quick use usage 3: many to many comparisons'
author: "Luis Ospina-Forero"
date: "23/06/2020"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ManyToManyComp}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---
  
```{r, include = FALSE}
  knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
  )
```
  
  # Introduction
  
  In some situations there is a need to compare multiple graphs among each other. For such situations, the `netdist` package contains some initial shortcut functions that perform such calculation and may automatically (Unix) incorporate parallel processing.
  
  This vignette does not show details about the Netdis and NetEmd network comparison methods or their variants, please check ["Default, simple and quick use usage 1: pairwise comparisons"](default_pairwise_usage.html) for these details. Instead, this vignette  shows the default usage of the shortcut functions for many-to-many comparisons.
  
  Please note that for high performance computation, this shortcut functions may or may not be ideal, particualry in terms of RAM consumption.
  
# Load required packages/libraries
```{r, packages, message= FALSE}
  # Load packages/libraries
library("netdist")
library("igraph")
library("pheatmap")
```

# Compare Multiple networks via NetEmd.

(Extracted from Wegner et al. (2017)): NetEmd is based on  the idea that the information encapsulated in the shape of the degree distribution and other network properties which reflect the topological organization of the network. From an abstract point of view, NetEmd views the shape of a distribution as a property that is invariant under linear deformations i.e$.$ translations and re-scalings of the axis.

## Networks being compared.

Generation of regular grid, ring and tree-like networks with 400 nodes and 1600 nodes. The plots of these networks illustrate clearly their structural differences.
```{r, netwokrs,fig.align='center'}
# Create networks
set.seed(3171)
gLat_1 <- graph.lattice(c(20,20)) 
gLat_2 <- graph.lattice(c(40,40))  
gRing_1 <- make_ring(20^2) 
gRing_2 <- make_ring(40^2)
gTree_1 <- igraph::as.undirected( make_tree(n = 20^2,children = 3) )
gTree_2 <- igraph::as.undirected( make_tree(n = 40^2,children = 3) )

plot(gLat_1,vertex.size=0.8,vertex.label=NA)
plot(gRing_1,vertex.size=0.8,vertex.label=NA)
plot(gTree_1,vertex.size=0.8,vertex.label=NA)
```

## NetEmd using orbit Counts

Orbit based NetEmd comparisons.
```{r, netemdorbits,fig.align='center'}
# NetEMD using subgraph counts.
glist <- list(Lat_1=gLat_1, Lat_2=gLat_2, Ring_1=gRing_1, Ring_2=gRing_1, Tree_1=gTree_1, Tree_2=gTree_2)

netemd_mat <- netemd_many_to_many(graphs = glist,smoothing_window_width = 1) #Use of smoothing window 1 is given for discrete integer distributions. If the network features are considered continuous variables smoothing_window_width equal to zero is recommended.

netemd_mat
```


Illustration of the multiple NetEmd comparisons based on orbit counts.
```{r,netemdorbitsPLOT,fig.align='center',fig.dim=c(8,8)}
# Creating matrices to plot the Network comparison
g_NETEMDcomp <- graph.edgelist(el = as.matrix(netemd_mat$comp_spec[,1:2]),directed = FALSE)
edge_attr(graph = g_NETEMDcomp,name = "NetEmdOrbits") <- netemd_mat$netemds
mat <- get.adjacency(graph = g_NETEMDcomp,type = "both",attr = "NetEmdOrbits",names = TRUE,sparse = FALSE)
vnames <- rownames(mat)
mat

#Network comparisons heatmap with Gold-Standard 1 
legend1 <- c(seq(min(mat),max(mat),length.out = 5))
levels1 <- round(legend1,digits = 2); levels1[length(levels1)]="NetEmd"
pheatmap::pheatmap(mat = as.dist(mat),cluster_rows = TRUE,cluster_cols = TRUE,clustering_method = "ward.D",angle_col=45,main = "NetEmd Orbit  Counts",labels_row = vnames,labels_col = vnames,display_numbers = TRUE,legend_breaks = legend1,legend_labels = levels1)
```

## NetEmd using the Laplacian and Normalized Laplacian Spectrum.
```{r, netemdspectrum,fig.align='center'}
# NetEMD using the Laplacian and normalized Laplacian Spectrum.
SPECT<-list()
#this step may take several minutes.
for(i in 1:length(glist)){ 
  Lapg <- igraph::laplacian_matrix(graph = glist[[i]],normalized = FALSE,sparse = FALSE)
  NLap <- igraph::laplacian_matrix(graph = glist[[i]],normalized = TRUE,sparse = FALSE)
  SPECT[[ names(glist)[i] ]] <- cbind(L.Spectra= eigen(Lapg)$values, NL.Spectra= eigen(NLap)$values) 
}

netemd_mat <- netemd_many_to_many(dhists = SPECT,smoothing_window_width = 0)#Use of smoothing window 1 is given for discrete integer distributions. If the network features are considered continuous variables smoothing_window_width equal to zero is recommended.

netemd_mat
```


Illustration of the multiple NetEmd comparisons based on the Laplacian and Normalized Laplacian spectra.
```{r,netemdspectrumPLOT,fig.align='center'}
# Creating matrices to plot the Network comparison
g_NETEMDcomp <- graph.edgelist(el = as.matrix(netemd_mat$comp_spec[,1:2]),directed = FALSE)
edge_attr(graph = g_NETEMDcomp,name = "NetEmdOrbits") <- netemd_mat$netemds
mat <- get.adjacency(graph = g_NETEMDcomp,type = "both",attr = "NetEmdOrbits",names = TRUE,sparse = FALSE)
vnames <- rownames(mat)
mat

#Network comparisons heatmap with Gold-Standard 1 
legend1 <- c(seq(min(mat),max(mat),length.out = 5))
levels1 <- round(legend1,digits = 2); levels1[length(levels1)]="NetEmd"
pheatmap::pheatmap(mat = as.dist(mat),cluster_rows = TRUE,cluster_cols = TRUE,clustering_method = "ward.D",angle_col=45,main = "NetEmd Spectra",labels_row = vnames,labels_col = vnames,display_numbers = TRUE,legend_breaks = legend1,legend_labels = levels1)
```


-------------------------

# Compare two networks via Netdis and its variants.
(Extracted from Ali et al. (2014)): Netdis counts small subgraphs  $w$ on $k$ nodes for all 2-step ego-networks, $k=3,4,5$. These counts are centred by subtracting the expected number of counts $E_w$. These centred counts of each network are then compared thus leading to the Netdis statistic.

## Using Netdis with a reference or gold-standard graph to obtain the expectations $E_w$.
The selection of a gold-standard graph as a substitute for a theoretical $E_w$ could be done when such graph is known to be a good proxy for $E_w$, or alternatively as a good reference point for the comparison. This netdis variant will focus on detecting discrepancies between the networks relative to the ego-network structure of the reference network / gold-standard given in $E_w$.


# Netdis using reference or goldstandard networks.

Generation of regular grid, ring and tree-like networks with 400 nodes and 1600 nodes. Reference graphs given by start-like graphs are used as illustration of reference graphs. The plots of these networks illustrate clearly their structural differences.
```{r,netdisgoldstandnetworks,fig.align='center'}
# Create lattice, Ring and Tree like networks of sizes 20^2 and 40^2.
# Create networks
set.seed(3171)
gLat_1 <- graph.lattice(c(20,20)) 
gLat_2 <- graph.lattice(c(40,40))  
gRing_1 <- make_ring(20^2) 
gRing_2 <- make_ring(40^2)
gTree_1 <- igraph::as.undirected( make_tree(n = 20^2,children = 3) )
gTree_2 <- igraph::as.undirected( make_tree(n = 40^2,children = 3) )

# Create a random graph to be used as a gold-standard
gst_1 <- graph.star(20^2,)
gst_2 <- graph.star(40^2)

plot(gst_1,vertex.size=0.8,vertex.label=NA)
```


Obtain the comparison via Netdis using each of the reference graph networks.
```{r,netdisgoldstand,fig.align='center'}
# Netdis using the goldstd_1 graph as gold-standard reference point
glist <- list(Lat_1=gLat_1, Lat_2=gLat_2, Ring_1=gRing_1, Ring_2=gRing_1, Tree_1=gTree_1, Tree_2=gTree_2)

netdis_mat_gst1 <- netdis_many_to_many(graphs = glist, ref_graph = gst_1)
netdis_mat_gst2 <- netdis_many_to_many(graphs = glist, ref_graph = gst_2)

netdis_mat_gst1

netdis_mat_gst2
```


```{r,netdisgoldstandPLOT,fig.align='center'}
# Creating matrices to plot the Network comparison with Gold-Standard 1 
g_NETDIScomp_1 <- graph.edgelist(el = as.matrix(netdis_mat_gst1$comp_spec[,1:2]),directed = FALSE)
edge_attr(graph = g_NETDIScomp_1,name = "Netdis4") <- netdis_mat_gst1$netdis[2,]
mat_gst1 <- get.adjacency(graph = g_NETDIScomp_1,type = "both",attr = "Netdis4",names = TRUE,sparse = FALSE)
vnames <- rownames(mat_gst1)
mat_gst1

# Creating matrices to plot the Network comparison with Gold-Standard 2 
g_NETDIScomp_2 <- graph.edgelist(el = as.matrix(netdis_mat_gst2$comp_spec[,1:2]),directed = FALSE)
edge_attr(graph = g_NETDIScomp_2,name = "Netdis4") <- netdis_mat_gst2$netdis[2,]
mat_gst2 <- get.adjacency(graph = g_NETDIScomp_2,type = "both",attr = "Netdis4",names = TRUE,sparse = FALSE)
vnames <- rownames(mat_gst2)
mat_gst2

#Network comparisons heatmap with Gold-Standard 1 
legend1 <- c(seq(min(mat_gst1),max(mat_gst1),length.out = 5))
levels1 <- round(legend1,digits = 2); levels1[length(levels1)]="Netdis"
pheatmap::pheatmap(mat = as.dist(mat_gst1),cluster_rows = TRUE,clustering_method = "ward.D",angle_col=45,main = "Netdis GoldStd-1",treeheight_row = 80,labels_row = vnames,labels_col = vnames,display_numbers = TRUE,legend_breaks = legend1,legend_labels = levels1)

#Network comparisons heatmap with Gold-Standard 2
legend2 <- c(seq(min(mat_gst2),max(mat_gst2),length.out = 5))
levels2 <- round(legend2,digits = 2); levels2[length(levels2)]="Netdis"
pheatmap::pheatmap(mat = as.dist(mat_gst2),cluster_rows = TRUE,clustering_method = "ward.D",angle_col=45,main = "Netdis GoldStd-2",treeheight_row = 80,labels_row = vnames,labels_col = vnames,display_numbers = TRUE,legend_breaks = legend2,legend_labels = levels2,kmeans_k = NA)
```


## Netdis-GP: Using a Geometric-Poisson approximation

This variant focuses on detecting more general and global discrepancies between the ego-network structures.

```{r, netdisGP}
# Netdis Geometric-Poisson comparisons
netdis_mat <- netdis_many_to_many(graphs = glist, ref_graph = NULL)
netdis_mat
```

```{r,netdisGPPLOT,fig.align='center'}
# Creating matrices to plot the Network comparison with Gold-Standard 1 
g_NETDIScomp_GP <- graph.edgelist(el = as.matrix(netdis_mat$comp_spec[,1:2]),directed = FALSE)
edge_attr(graph = g_NETDIScomp_GP,name = "Netdis4") <- netdis_mat$netdis[2,]
mat_GP <- get.adjacency(graph = g_NETDIScomp_GP,type = "both",attr = "Netdis4",names = TRUE,sparse = FALSE)
vnames <- rownames(mat_GP)
mat_GP

#Network comparisons heatmap with Gold-Standard 1 
legend1 <- c(seq(min(mat_GP),max(mat_GP),length.out = 5))
levels1 <- round(legend1,digits = 2); levels1[length(levels1)]="Netdis"
pheatmap::pheatmap(mat = as.dist(mat_GP),cluster_rows = TRUE,clustering_method = "ward.D",angle_col=45,main = "Netdis-GP",treeheight_row = 80,labels_row = vnames,labels_col = vnames,display_numbers = TRUE,legend_breaks = legend1,legend_labels = levels1)
```

## Using Netdis with no expectation ($E_w=0$)
Comparing the networks via their observed ego counts without centering them, (equivalent to using expectation equal to zero). This variant thus focuses on detecting small discrepancies between the networks.



```{r, netdiszero}
# Netdis Geometric-Poisson comparisons
netdis_mat <- netdis_many_to_many(graphs = glist, ref_graph = 0)
netdis_mat
```

```{r,netdiszeroPLOT,fig.align='center'}
# Creating matrices to plot the Network comparison with Gold-Standard 1 
g_NETDIScomp_zero <- graph.edgelist(el = as.matrix(netdis_mat$comp_spec[,1:2]),directed = FALSE)
edge_attr(graph = g_NETDIScomp_zero,name = "Netdis4") <- netdis_mat$netdis[2,]
mat_zero <- get.adjacency(graph = g_NETDIScomp_zero,type = "both",attr = "Netdis4",names = TRUE,sparse = FALSE)
vnames <- rownames(mat_zero)
mat_zero

#Network comparisons heatmap with Gold-Standard 1 
legend1 <- c(seq(min(mat_zero),max(mat_zero),length.out = 5))
levels1 <- round(legend1,digits = 2); levels1[length(levels1)]="Netdis"
pheatmap::pheatmap(mat = as.dist(mat_zero),cluster_rows = TRUE,clustering_method = "ward.D",angle_col=45,main = "Netdis-zero",treeheight_row = 80,labels_row = vnames,labels_col = vnames,display_numbers = TRUE,legend_breaks = legend1,legend_labels = levels1)
```

-------------------------


# Bibliography

* W. Ali, T. Rito, G. Reinert, F. Sun, and C. M. Deane. Alignment-free protein interaction network comparison. Bioinformatics, 30:i430–i437, 2014.

* L. Ospina-Forero, C. M. Deane, and G. Reinert. Assessment of model fit via network comparison methods based on subgraph counts. Journal of Complex Networks, page cny017, August 2018.

* A. E. Wegner, L. Ospina-Forero, R. E. Gaunt, C. M. Deane, and G. Reinert. Identifying networks with common organizational principles. Journal of Complex networks, 2017.

* F. Picard, J.-J. Daudin, M. Koskas, S. Schbath, and S. Robin. Assessing the exceptionality of network motifs. Journal of Computational Biology, 15(1):1–20, 2008.