<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simple and quick (default) usage 1: pairwise comparisons • netdist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simple and quick (default) usage 1: pairwise comparisons">
<meta property="og:description" content="netdist">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">netdist</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ManyToMany.html">Simple and quick (default) usage 3: many to many comparisons</a>
    </li>
    <li>
      <a href="../articles/NetEmdTimeOrdering.html">NetEmd: World trade</a>
    </li>
    <li>
      <a href="../articles/NetdisGPStepByStep.html">Netdis-GP: step by step</a>
    </li>
    <li>
      <a href="../articles/NetdisStepByStep.html">Netdis: step by step</a>
    </li>
    <li>
      <a href="../articles/NewNetdisCustomisations.html">Netdis: customizations</a>
    </li>
    <li>
      <a href="../articles/PreComputedProps.html">Simple and quick (default) usage 2: pre-computed properties</a>
    </li>
    <li>
      <a href="../articles/V-Menu.html">Netdis Vignette's Menu</a>
    </li>
    <li>
      <a href="../articles/default_pairwise_usage.html">Simple and quick (default) usage 1: pairwise comparisons</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_dis.html">Netdis: Protein interaction networks</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_emd.html">Dendrogram example for NetEMD</a>
    </li>
    <li>
      <a href="../articles/netdis_customisations.html">Usage of netdis with binning and expected counts customisations.</a>
    </li>
    <li>
      <a href="../articles/netdis_pairwise_comparisons.html">Usage of netdis interfaces for different pairwise comparison options.</a>
    </li>
    <li>
      <a href="../articles/quickstart_netdis_2graphs.html">Quick start guide for Netdis - 2 graphs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simple and quick (default) usage 1: pairwise comparisons</h1>
            
            <h4 data-toc-skip class="date"><code>10-06-2020</code></h4>
      
      
      <div class="hidden name"><code>default_pairwise_usage.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>netdist</code> package currently considers two broad methodologies for network comparison, namely Netdis and NetEmd. Netdis considers multiple variants (via background expectations) to capture the dissimilarity between the local structure of networks exhibited by the occurrence of small subgraphs on 3, 4 and 5 nodes. NetEmd is also a method to capture the dissimilarity between networks using subgraph counts, but it has in addition been defined for any type of network features; for example eigen distributions. The variants of Netdis are controlled by the input selected for the background expectations, whereas the variants of NetEmd are controlled indirectly by the user by the selection of the network features being compared via NetEmd (by default this package uses subgraph counts as features).</p>
<p>The following shows a quick introduction to the simplest functions of the package, and to some of the variants of Netdis and NetEmd.</p>
<p>For other vignettes in this package see the <a href="V-Menu.html">“Menu”</a>.</p>
</div>
<div class="section level2">
<h2 id="load-required-packageslibraries">Load required packages/libraries<a class="anchor" aria-label="anchor" href="#load-required-packageslibraries"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load packages/libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"netdist"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://igraph.org" class="external-link">"igraph"</a></span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="load-protein-interaction-graphs-included-in-the-netdist-package">Load protein interaction graphs included in the netdist package<a class="anchor" aria-label="anchor" href="#load-protein-interaction-graphs-included-in-the-netdist-package"></a>
</h3>
<p>The <code>netdist</code> package also includes examples of a few real networks. These are protein interaction networks (PPI) of a few <em>Herpes</em> virus and of the bacteria <em>Escherichia Coli</em> . In these networks, each node represents a protein and each link represents an interaction between proteins. See <code><a href="../reference/virusppi.html">help(virusppi)</a></code>.</p>
<p>Although the <code>virusppi</code> list of PPI networks is loaded along with the <code>netdist</code> package, the following code shows how to read a network data from a file in disk:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set source directory for Virus protein-protein interaction edge files stored in the netdist package.</span>
<span class="va">source_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"VRPINS"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"netdist"</span><span class="op">)</span>

<span class="co"># Load query graphs as undirected igraph objects, with no loops, multiple edges or degree zero nodes.</span>
<span class="va">graph_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"EBV.txt"</span><span class="op">)</span>,
                             format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>

<span class="va">graph_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"ECL.txt"</span><span class="op">)</span>,
                             format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>

<span class="co"># Herpes virus EBV protein-protein interaction graph with 60 nodes and 208 edges.</span>
<span class="va">graph_1</span>
<span class="co">#&gt; IGRAPH 24d036b UN-- 60 208 -- </span>
<span class="co">#&gt; + attr: name (v/c)</span>
<span class="co">#&gt; + edges from 24d036b (vertex names):</span>
<span class="co">#&gt;  [1] A73  --BALF3 A73  --BARF0 A73  --BBLF2 A73  --BDRF1 A73  --BFRF4</span>
<span class="co">#&gt;  [6] A73  --BGLF2 A73  --BGLF3 A73  --BGLF5 A73  --BLLF2 A73  --BTRF1</span>
<span class="co">#&gt; [11] BALF3--BBLF2 BALF3--BDRF1 BALF3--BFRF4 BALF3--BGLF5 BALF3--BTRF1</span>
<span class="co">#&gt; [16] BALF3--BALF1 BALF3--BALF2 BALF3--BORF1 BALF3--BALF4 BALF3--BFLF2</span>
<span class="co">#&gt; [21] BALF3--BPLF1 BALF3--BALF5 BALF3--BBLF4 BALF3--BDLF2 BALF3--BdRF1</span>
<span class="co">#&gt; [26] BALF3--BERF3 BALF3--BHRF1 BALF3--LMP2A BARF0--BBLF2 BARF0--BFRF4</span>
<span class="co">#&gt; [31] BARF0--BSRF1 BARF0--BALF4 BARF0--BPLF1 BARF0--BALF5 BARF0--BDLF2</span>
<span class="co">#&gt; [36] BARF0--BdRF1 BARF0--BERF3 BARF0--BGLF1 BARF0--LMP2A BBLF2--BDRF1</span>
<span class="co">#&gt; + ... omitted several edges</span>
<span class="co">#Note this graph is the same as</span>
<span class="co"># virusppi$EBV</span>

<span class="co"># Herpes virus ECL protein-protein interaction graph with 1941 nodes and 3989 edges.</span>
<span class="va">graph_2</span>
<span class="co">#&gt; IGRAPH 900e8b6 UN-- 1941 3989 -- </span>
<span class="co">#&gt; + attr: name (v/c)</span>
<span class="co">#&gt; + edges from 900e8b6 (vertex names):</span>
<span class="co">#&gt;  [1] B1882--B1888 B1882--B1945 B1882--B1946 B1882--B1886 B1882--B1887</span>
<span class="co">#&gt;  [6] B1882--B1939 B1882--B1938 B1882--B1884 B1882--B1883 B1882--B3210</span>
<span class="co">#&gt; [11] B1882--B1881 B1882--B4355 B1882--B1922 B1882--B1890 B1882--B1889</span>
<span class="co">#&gt; [16] B1888--B1886 B1888--B1887 B1888--B1884 B1888--B1883 B1888--B1881</span>
<span class="co">#&gt; [21] B1888--B4355 B1888--B1890 B1888--B1889 B1888--B1421 B1888--B3072</span>
<span class="co">#&gt; [26] B1888--B1885 B0728--B0729 B0728--B0724 B0728--B0726 B0728--B0727</span>
<span class="co">#&gt; [31] B0729--B0724 B0729--B3734 B0729--B0726 B0729--B0727 B0729--B0720</span>
<span class="co">#&gt; [36] B0729--B3236 B1812--B3360 B1812--B1260 B1812--B1261 B1812--B1263</span>
<span class="co">#&gt; + ... omitted several edges</span>
<span class="co">#Note this graph is the same as</span>
<span class="co"># virusppi$ECL</span>

<span class="co">#A simple visualization of the graphs.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_1</span>,vertex.size<span class="op">=</span><span class="fl">4</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="default_pairwise_usage_files/figure-html/graphs-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_2</span>,vertex.size<span class="op">=</span><span class="fl">4</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="default_pairwise_usage_files/figure-html/graphs-2.png" width="576" style="display: block; margin: auto;"></p>
<p>Other networks loaded in this package are discussed in <a href="NetEmdTimeOrdering.html">“NetEmd: World trade networks”</a>. You can also see <code><a href="../reference/virusppi.html">?virusppi</a></code> and <code><a href="../reference/worldtradesub.html">?worldtradesub</a></code>.</p>
<p>The next two sections will provide a basic introduction and usage of both NetEmd and Netdis.</p>
</div>
</div>
<div class="section level2">
<h2 id="compare-two-networks-via-netemd">Compare two networks via NetEmd<a class="anchor" aria-label="anchor" href="#compare-two-networks-via-netemd"></a>
</h2>
<div class="section level3">
<h3 id="what-is-netemd">What is NetEmd?<a class="anchor" aria-label="anchor" href="#what-is-netemd"></a>
</h3>
<p>(Extracted from Wegner et al. (2017)): NetEmd is based on the idea that the information encapsulated in the shape of the degree distribution and other network properties reflects the topological organization of the network. From an abstract point of view, NetEmd views the shape of a distribution as a property that is invariant under linear deformations i.e<span class="math inline">\(.\)</span> translations and re-scalings of the axis.</p>
<p>Based on the previous ideas NetEmd uses the following measure between distributions <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> that are supported on <span class="math inline">\(\mathbb{R}\)</span> and have non-zero, finite variances: <span class="math display">\[\begin{equation}\label{emdmet}
EMD^*(p,q)=\mathrm{inf}_{c\in\mathbb{R}}\left( EMD\big(\tilde{p}(\cdot+c),\tilde{q}(\cdot)\big)\right),
\end{equation}\]</span> where <span class="math inline">\(EMD\)</span> is the earth mover’s distance and <span class="math inline">\(\tilde{p}\)</span> and <span class="math inline">\(\tilde{q}\)</span> are the distributions obtained by rescaling <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> to have variance 1. More precisely, <span class="math inline">\(\tilde{p}\)</span> is the distribution obtained from <span class="math inline">\(p\)</span> by the transformation <span class="math inline">\(x\rightarrow \frac{x}{\sigma(p)}\)</span>, where <span class="math inline">\(\sigma(p)\)</span> is the standard deviation of <span class="math inline">\(p\)</span>. For probability distributions <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> with support in <span class="math inline">\(\mathbb{R}\)</span> and bounded absolute first moment, the <span class="math inline">\(EMD\)</span> between <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> is given by <span class="math inline">\(EMD(p,q)=\int_{-\infty}^\infty|F(x)-G(x)|\,\mathrm{d}x\)</span>, where <span class="math inline">\(F\)</span> and <span class="math inline">\(G\)</span> are the cumulative distribution functions of <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> respectively.</p>
<p>Now, for two networks <span class="math inline">\(G\)</span> and <span class="math inline">\(G'\)</span> and for a given set <span class="math inline">\(T=\{t_1,t_2,...,t_m\}\)</span> of network features, the <span class="math inline">\(NetEmd\)</span> measure corresponding to <span class="math inline">\(T\)</span> is: <span class="math display">\[\begin{equation}\label{eq:def_netemd}
NetEmd_T(G,G')=\frac{1}{m}\sum_{j=1}^{m} NetEmd_{t_j} (G,G'),
\end{equation}\]</span> where <span class="math display">\[\begin{equation}
NetEmd_{t_i} (G,G')=EMD^*(p_{t_i}(G),p_{t_i}(G')),
\end{equation}\]</span> and where <span class="math inline">\(p_{t_i}(G)\)</span> and <span class="math inline">\(p_{t_i}(G')\)</span> are the distributions of <span class="math inline">\({t_i}\)</span> on <span class="math inline">\(G\)</span> and <span class="math inline">\(G'\)</span> respectively. <span class="math inline">\(NetEmd_{t_i}\)</span> can be shown to be a pseudometric between graphs for any feature <span class="math inline">\(t\)</span>, that is it is non-negative, symmetric and satisfies the triangle inequality.</p>
</div>
<div class="section level3">
<h3 id="comparing-two-graphs-with-netemd">Comparing two graphs with NetEmd<a class="anchor" aria-label="anchor" href="#comparing-two-graphs-with-netemd"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set source directory for Virus protein-protein interaction network edge files stored in the netdist package.</span>
<span class="va">source_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"VRPINS"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"netdist"</span><span class="op">)</span>

<span class="co"># Load query graphs as igraph objects</span>
<span class="co"># Herpes virus EBV protein-protein interaction graph with 60 nodes and 208 edges.</span>
<span class="va">graph_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"EBV.txt"</span><span class="op">)</span>,
                             format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>

<span class="co"># Herpes virus ECL protein-protein interaction graph with 1941 nodes and 3989 edges.</span>
<span class="va">graph_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"ECL.txt"</span><span class="op">)</span>,
                             format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>

<span class="co"># One to one NetEmd comparison.</span>
<span class="fu"><a href="../reference/netemd_one_to_one.html">netemd_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span><span class="va">graph_1</span>,graph_2<span class="op">=</span><span class="va">graph_2</span>,feature_type<span class="op">=</span><span class="st">"orbit"</span>,smoothing_window_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="co">#Use of smoothing window 1 is given for discrete integer distributions. If the network features are considered continuous variables smoothing_window_width equal to zero is recommended.</span>
<span class="co">#&gt; [1] 0.5163894</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparing-two-graphs-with-netemd-via-their-laplacian-spectrum">Comparing two graphs with NetEmd via their Laplacian spectrum<a class="anchor" aria-label="anchor" href="#comparing-two-graphs-with-netemd-via-their-laplacian-spectrum"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Laplacian</span>
<span class="va">Lapg_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_1</span>,normalized <span class="op">=</span> <span class="cn">FALSE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">Lapg_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_2</span>,normalized <span class="op">=</span> <span class="cn">FALSE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#Normalized Laplacian</span>
<span class="va">NLapg_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_1</span>,normalized <span class="op">=</span> <span class="cn">TRUE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">NLapg_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_2</span>,normalized <span class="op">=</span> <span class="cn">TRUE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#Spectra (this may take a couple of minutes).</span>
<span class="va">props_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>L.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">Lapg_1</span><span class="op">)</span><span class="op">$</span><span class="va">values</span>, NL.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">NLapg_1</span><span class="op">)</span><span class="op">$</span><span class="va">values</span><span class="op">)</span> 
<span class="va">props_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>L.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">Lapg_2</span><span class="op">)</span><span class="op">$</span><span class="va">values</span>, NL.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">NLapg_2</span><span class="op">)</span><span class="op">$</span><span class="va">values</span><span class="op">)</span> 

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">props_1</span>,n<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;      L.Spectra NL.Spectra</span>
<span class="co">#&gt; [1,]  28.20519   1.761482</span>
<span class="co">#&gt; [2,]  23.52699   1.659654</span>
<span class="co">#&gt; [3,]  20.58109   1.620187</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">props_2</span>,n<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;      L.Spectra NL.Spectra</span>
<span class="co">#&gt; [1,]  58.14333          2</span>
<span class="co">#&gt; [2,]  42.13825          2</span>
<span class="co">#&gt; [3,]  33.30286          2</span>

<span class="fu"><a href="../reference/netemd_one_to_one.html">netemd_one_to_one</a></span><span class="op">(</span>dhists_1 <span class="op">=</span> <span class="va">props_1</span>,dhists_2 <span class="op">=</span> <span class="va">props_2</span>,smoothing_window_width <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="co">#If the network features are considered continuous variables smoothing_window_width equal to zero is recommended.</span>
<span class="co">#&gt; [1] 0.1818771</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="compare-two-networks-via-netdis-and-its-variants">Compare two networks via Netdis and its variants<a class="anchor" aria-label="anchor" href="#compare-two-networks-via-netdis-and-its-variants"></a>
</h2>
<div class="section level3">
<h3 id="what-is-netdis">What is Netdis?<a class="anchor" aria-label="anchor" href="#what-is-netdis"></a>
</h3>
<p>(Extracted from Ali et al. (2014)): Netdis counts small subgraphs <span class="math inline">\(w\)</span> on <span class="math inline">\(k\)</span> nodes for all 2-step ego-networks, <span class="math inline">\(k=3,4,5\)</span>. These counts are centred by subtracting the expected number of counts <span class="math inline">\(E_w\)</span>. These centred counts of each network are then compared thus leading to the Netdis statistic.</p>
<p>Netdis is constructed as follows:</p>
<p>Let <span class="math inline">\(N_{w,i}(G)\)</span> be the number of induced occurrences of small graphs <span class="math inline">\(w\)</span> in the 2-step ego network of vertex <span class="math inline">\(i\)</span>. Now, bin all 2-step ego-networks of network <span class="math inline">\(G\)</span> according to their network density. Let <span class="math inline">\(E_w(G,\rho)\)</span> be the expected number of occurrences of <span class="math inline">\(w\)</span> in an ego-network whose density falls in density bin <span class="math inline">\(\rho\)</span>. For a given network <span class="math inline">\(G\)</span> compute the centred subgraph counts as <span class="math display">\[
S_w(G)=\sum\limits_{i }{\bigg (N_{w,i}(G)- E_w(G, \rho(i)) \bigg )},
\]</span> where <span class="math inline">\(i\)</span> is a node in <span class="math inline">\(G\)</span> and <span class="math inline">\(\rho(i)\)</span> the density bin of the 2-step ego-network of node <span class="math inline">\(i\)</span>.</p>
<p>Now, to compare networks <span class="math inline">\(G_1\)</span> and <span class="math inline">\(G_2\)</span>, set <span class="math display">\[
\displaystyle
netD_2^S(k) = \tfrac{1}{ \sqrt{ M(k)} } \sum\limits_{w \in A(k)}
\bigg ({ \tfrac{S_w(G_1)  S_w(G_2)} {\sqrt{S_w(G_1)^2 + S_w(G_2)^2}}
}\bigg ), \quad k=3,4, 5, 
\]</span> where <span class="math inline">\(A(k)\)</span> is the set of connected subgraphs of size <span class="math inline">\(k\)</span>, and where <span class="math inline">\(M(k)\)</span> is a normalising constant so that <span class="math inline">\(netD_2^S(k)\in[-1,1]\)</span>. <span class="math inline">\(M(k)\)</span> is equal to <span class="math display">\[
M(k)  = \sum\limits_{w \in A(k)}
\left( \tfrac{ S_w(G_1)^2 }{\sqrt{S_w(G_1)^2 + S_w(G_2)^2}} \right)
\sum\limits_{w \in A(k)}
\left(\tfrac{ S_w(G_2)^2 } {\sqrt{S_w(G_1)^2 + S_w(G_2)^2}}
\right)
.
\]</span> The corresponding Netdis statistic is defined as <span class="math display">\[Netdis(k)=netd_2^S(k)=\tfrac{1}{2}(1-netD_2^S(k)) \in [0,1].\]</span> Small values of Netdis suggest higher `similarity’ between the networks. By default Netdis uses subgraphs on <span class="math inline">\(k=4\)</span> nodes.</p>
</div>
<div class="section level3">
<h3 id="using-netdis-with-a-gold-standard-graph-to-obtain-e_w">Using Netdis with a gold-standard graph to obtain <span class="math inline">\(E_w\)</span><a class="anchor" aria-label="anchor" href="#using-netdis-with-a-gold-standard-graph-to-obtain-e_w"></a>
</h3>
<p>The selection of a gold-standard graph as a substitute for <span class="math inline">\(E_w\)</span> could be done when such graph is known to be a good proxy for <span class="math inline">\(E_w\)</span>, or alternatively as a good reference point for the comparison. This option will focus on detecting discrepancies between the networks relative to the ego-network structure of the reference network / gold-standard graph and which is summarized in <span class="math inline">\(E_w\)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Lattice graphs to be used as a gold-standard reference point</span>
<span class="va">goldstd_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span> <span class="co">#Graph with 8^2 nodes</span>
<span class="va">goldstd_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">44</span>,<span class="fl">44</span><span class="op">)</span><span class="op">)</span> <span class="co">#Graph with 44^2 nodes</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">goldstd_1</span>,vertex.size<span class="op">=</span><span class="fl">4</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="default_pairwise_usage_files/figure-html/netdisgoldstand-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">goldstd_2</span>,vertex.size<span class="op">=</span><span class="fl">4</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="default_pairwise_usage_files/figure-html/netdisgoldstand-2.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="co"># Netdis using the goldstd_1 graph as gold-standard reference point</span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="va">goldstd_1</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.1422771 0.2517043</span>

<span class="co"># Netdis using the goldstd_2 graph as gold-standard reference point</span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="va">goldstd_2</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.1401654 0.2505384</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="netdis-gp-using-a-geometric-poisson-approximation">Netdis-GP: Using a Geometric-Poisson approximation<a class="anchor" aria-label="anchor" href="#netdis-gp-using-a-geometric-poisson-approximation"></a>
</h3>
<p>(Extracted from Ospina-Forero et al. (2018)): Instead of considering an approximation based on an observed gold-standard network whose selection may be difficult, <span class="math inline">\(E_w\)</span> is computed independently for each graph, based on a Geometric-Poisson (GP) approximation of the distribution of the number of occurrences of subgraph <span class="math inline">\(w\)</span>. It assumes that <span class="math inline">\(N_{w,i} \sim GP(\lambda^{\rho(i)}_k, \theta^{\rho(i)}_w)\)</span>, where <span class="math inline">\(\lambda^{\rho(i)}_k\)</span> is the Poisson parameter indexed by the size of subgraph <span class="math inline">\(w\)</span> and the density bin <span class="math inline">\(\rho(i)\)</span>; and where <span class="math inline">\(\theta^{\rho(i)}_w\)</span> is the geometric parameter indexed by subgraph <span class="math inline">\(w\)</span> and density bin <span class="math inline">\(\rho(i)\)</span>. <span class="math inline">\(E_w(G, \rho(i))\)</span> is taken as the mean of the GP approximation, i.e. <span class="math inline">\(\lambda^{\rho(i)}_k/\theta^{\rho(i)}_w\)</span>.</p>
<p>As <span class="math inline">\(\lambda^{\rho(i)}_k\)</span> and <span class="math inline">\(\theta^{\rho(i)}_w\)</span> are not known, they are estimated as follows: Let <span class="math inline">\(x_{w,d}^j\)</span> be the number of subgraphs <span class="math inline">\(w\)</span> on the 2-step ego-network <span class="math inline">\(j\)</span> of density bin <span class="math inline">\(d\)</span>, and let <span class="math display">\[
\bar{X}_{w,d}=\frac{1}{q} \sum_{j=1}^q x_{w,d}^j, \qquad V^2_{w,d}=\frac{1}{q-1} \sum_{j=1}^q (x_{w,d}^j - \bar{X}_{w,d})^2
,
\]</span> where <span class="math inline">\(q\)</span> is the number of ego-networks in density bin <span class="math inline">\(d\)</span>. Then, <span class="math display">\[
\hat{\lambda}^{d}_{k}= \frac{1}{l} \sum_{h \in A(k)} \frac{2 (\bar{X}_{h,d})^2}{V^2_{h,d}+\bar{X}_{h,d}} , \qquad \hat{\theta}^{d}_w= \frac{2\bar{X}_{w,d}}{V^2_{w,d}+\bar{X}_{w,d}},
\]</span> where <span class="math inline">\(l\)</span> is the number of connected subgraphs of size <span class="math inline">\(k\)</span>, for example, <span class="math inline">\(l=6\)</span> for <span class="math inline">\(k=4\)</span>. These estimators are based on the moment estimators of a GP random variable and the proposal made by (Picard et al.(2008)), where the total count of each individual subgraph could be thought as the sum of the total subgraph counts over multiple ``clumps’’ of edges that appear across the network.</p>
<p>This variant focuses on detecting more meso-level discrepancies between the ego-network structures.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Netdis using the Geometric-Poisson approximation as a way to obtain background expectations. </span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.8822527 0.1892716</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-netdis-with-no-expectation-e_w0">Using Netdis with no expectation (<span class="math inline">\(E_w=0\)</span>)<a class="anchor" aria-label="anchor" href="#using-netdis-with-no-expectation-e_w0"></a>
</h3>
<p>Comparing the networks via their observed ego counts without centring them, (equivalent to using expectation equal to zero). This variant thus focus on detecting small discrepancies between the networks.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Netdis using no expectations (or equivalently, expectation equal to zero).</span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt;    netdis3    netdis4 </span>
<span class="co">#&gt; 0.00761545 0.02106628</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<ul>
<li><p>W. Ali, T. Rito, G. Reinert, F. Sun, and C. M. Deane. Alignment-free protein interaction network comparison. Bioinformatics, 30:i430–i437, 2014.</p></li>
<li><p>A. E. Wegner, L. Ospina-Forero, R. E. Gaunt, C. M. Deane, and G. Reinert. Identifying networks with common organizational principles. Journal of Complex networks, 2017.</p></li>
<li><p>L. Ospina-Forero, C. M. Deane, and G. Reinert. Assessment of model fit via network comparison methods based on subgraph counts. Journal of Complex Networks, page cny017, August 2018.</p></li>
<li><p>F. Picard, J.-J. Daudin, M. Koskas, S. Schbath, and S. Robin. Assessing the exceptionality of network motifs. Journal of Computational Biology, 15(1):1–20, 2008.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin O’Reilly.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
