<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Netdis: step by step • netdist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Netdis: step by step">
<meta property="og:description" content="netdist">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">netdist</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ManyToMany.html">Simple and quick (default) usage 3: many to many comparisons</a>
    </li>
    <li>
      <a href="../articles/NetEmdTimeOrdering.html">NetEmd: World trade</a>
    </li>
    <li>
      <a href="../articles/NetdisGPStepByStep.html">Netdis-GP: step by step</a>
    </li>
    <li>
      <a href="../articles/NetdisStepByStep.html">Netdis: step by step</a>
    </li>
    <li>
      <a href="../articles/NewNetdisCustomisations.html">Netdis: customizations</a>
    </li>
    <li>
      <a href="../articles/PreComputedProps.html">Simple and quick (default) usage 2: pre-computed properties</a>
    </li>
    <li>
      <a href="../articles/V-Menu.html">Netdis Vignette's Menu</a>
    </li>
    <li>
      <a href="../articles/default_pairwise_usage.html">Simple and quick (default) usage 1: pairwise comparisons</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_dis.html">Netdis: Protein interaction networks</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_emd.html">Dendrogram example for NetEMD</a>
    </li>
    <li>
      <a href="../articles/netdis_customisations.html">Usage of netdis with binning and expected counts customisations.</a>
    </li>
    <li>
      <a href="../articles/netdis_pairwise_comparisons.html">Usage of netdis interfaces for different pairwise comparison options.</a>
    </li>
    <li>
      <a href="../articles/quickstart_netdis_2graphs.html">Quick start guide for Netdis - 2 graphs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Netdis: step by step</h1>
            
            <h4 data-toc-skip class="date">2022-06-06</h4>
      
      
      <div class="hidden name"><code>NetdisStepByStep.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette explains some inner calculations of Netdis which the user does not need to go through, however, we explain them here for those users that may want to use Netdis in a modular fashion.</p>
<p>For a simple Netdis function call see instead <a href="default_pairwise_usage.html">“Simple and quick (default) usage 1: pairwise comparisons”</a>.</p>
<p>For other vignettes in this package see the <a href="V-Menu.html">“Menu”</a>.</p>
</div>
<div class="section level2">
<h2 id="what-is-netdis">What is Netdis?<a class="anchor" aria-label="anchor" href="#what-is-netdis"></a>
</h2>
<p>(Extracted from Ali et al. (2014)): Netdis counts small subgraphs <span class="math inline">\(w\)</span> on <span class="math inline">\(k\)</span> nodes for all 2-step ego-networks, <span class="math inline">\(k=3,4,5\)</span>. These counts are centred by subtracting the expected number of counts <span class="math inline">\(E_w\)</span>. These centred counts of each network are then compared among one another thus leading to the Netdis statistic.</p>
<p>The selection of a gold-standard graph as a substitute for <span class="math inline">\(E_w\)</span> can be done when such graph is known to be a good proxy for <span class="math inline">\(E_w\)</span>, or alternatively as a good reference point for the comparison. This option will focus on detecting discrepancies between the networks relative to the ego-network structure of the reference network / gold-standard summarized in <span class="math inline">\(E_w\)</span>.</p>
<div class="section level3">
<h3 id="netdis-is-defined-as-follows">Netdis is defined as follows:<a class="anchor" aria-label="anchor" href="#netdis-is-defined-as-follows"></a>
</h3>
<p>Let <span class="math inline">\(N_{w,i}(G)\)</span> be the number of induced occurrences of small graphs <span class="math inline">\(w\)</span> in the 2-step ego-network of vertex <span class="math inline">\(i\)</span>. Now, bin all 2-step ego-networks of network <span class="math inline">\(G\)</span> according to their network density. Let <span class="math inline">\(E_w(G,\rho)\)</span> be the expected number of occurrences of <span class="math inline">\(w\)</span> in an ego-network whose density falls in density bin <span class="math inline">\(\rho\)</span>. For a given network <span class="math inline">\(G\)</span> compute the centred subgraph counts as <span class="math display">\[
S_w(G)=\sum\limits_{i }{\bigg (N_{w,i}(G)- E_w(G, \rho(i)) \bigg )},
\]</span> where <span class="math inline">\(i\)</span> is a node in <span class="math inline">\(G\)</span> and <span class="math inline">\(\rho(i)\)</span> the density bin of the 2-step ego-network of node <span class="math inline">\(i\)</span>.</p>
<p>Now, to compare networks <span class="math inline">\(G_1\)</span> and <span class="math inline">\(G_2\)</span>, set <span class="math display">\[
\displaystyle
netD_2^S(k) = \tfrac{1}{ \sqrt{ M(k)} } \sum\limits_{w \in A(k)}
\bigg ({ \tfrac{S_w(G_1)  S_w(G_2)} {\sqrt{S_w(G_1)^2 + S_w(G_2)^2}}
}\bigg ), \quad k=3,4, 5, 
\]</span> where <span class="math inline">\(A(k)\)</span> is the set of connected subgraphs of size <span class="math inline">\(k\)</span>, and where <span class="math inline">\(M(k)\)</span> is a normalising constant so that <span class="math inline">\(netD_2^S(k)\in[-1,1]\)</span>. <span class="math inline">\(M(k)\)</span> is equal to <span class="math display">\[
M(k)  = \sum\limits_{w \in A(k)}
\left( \tfrac{ S_w(G_1)^2 }{\sqrt{S_w(G_1)^2 + S_w(G_2)^2}} \right)
\sum\limits_{w \in A(k)}
\left(\tfrac{ S_w(G_2)^2 } {\sqrt{S_w(G_1)^2 + S_w(G_2)^2}}
\right)
.
\]</span> The corresponding Netdis statistic is defined as <span class="math display">\[Netdis(k)=netd_2^S(k)=\tfrac{1}{2}(1-netD_2^S(k)) \in [0,1].\]</span> Small values of Netdis suggest higher `similarity’ between the networks. By default Netdis uses subgraphs on <span class="math inline">\(k=4\)</span> nodes.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="netdis-step-by-step">Netdis step by step<a class="anchor" aria-label="anchor" href="#netdis-step-by-step"></a>
</h2>
<div class="section level3">
<h3 id="load-required-libraries">Load required libraries<a class="anchor" aria-label="anchor" href="#load-required-libraries"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"netdist"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://igraph.org" class="external-link">"igraph"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://purrr.tidyverse.org" class="external-link">"purrr"</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="networks-being-compared">Networks being compared<a class="anchor" aria-label="anchor" href="#networks-being-compared"></a>
</h3>
<p>Generation of tree-like networks with 400 nodes and 1600 nodes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create networks</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">34</span><span class="op">)</span>
<span class="va">gTree_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/as.directed.html" class="external-link">as.undirected</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_tree.html" class="external-link">make_tree</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">20</span><span class="op">^</span><span class="fl">2</span>,children <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">)</span>
<span class="va">gTree_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/as.directed.html" class="external-link">as.undirected</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_tree.html" class="external-link">make_tree</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">40</span><span class="op">^</span><span class="fl">2</span>,children <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gTree_1</span>,vertex.size<span class="op">=</span><span class="fl">0.8</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="NetdisStepByStep_files/figure-html/netwokrs-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gTree_2</span>,vertex.size<span class="op">=</span><span class="fl">0.8</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="NetdisStepByStep_files/figure-html/netwokrs-2.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="set-netdis-parameters">Set Netdis parameters<a class="anchor" aria-label="anchor" href="#set-netdis-parameters"></a>
</h3>
<p>Netdis uses some mostly internal parameters that define the size of the subgraphs/graphlets that are going to be used, as well as the length of the ego neighbourhood, the minimum size considered for the resulting ego-networks, and finally some parameters that control an ego-network binning function. By default, Netdis considers the following setup:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Maximum subgraph size to calculate counts and netdis statistic for</span>
<span class="va">max_subgraph_size</span> <span class="op">&lt;-</span> <span class="fl">4</span>

<span class="co"># Ego-network neighbourhood size</span>
<span class="va">neighbourhood_size</span> <span class="op">&lt;-</span> <span class="fl">2</span>

<span class="co"># Minimum size of ego-networks to consider</span>
<span class="va">min_ego_nodes</span> <span class="op">&lt;-</span> <span class="fl">3</span>
<span class="va">min_ego_edges</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="co"># Ego-network density binning parameters. Here, the minimum number of ego-networks per bin and the starting number of bins</span>
<span class="va">min_bin_count</span> <span class="op">&lt;-</span> <span class="fl">5</span>
<span class="va">num_bins</span> <span class="op">&lt;-</span> <span class="fl">100</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="obtain-list-of-ego-networks">Obtain list of ego-networks<a class="anchor" aria-label="anchor" href="#obtain-list-of-ego-networks"></a>
</h3>
<p>One of the first steps in Netdis is the extraction of all ego-networks in each of the query networks:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get ego-networks for query graphs</span>
<span class="va">ego_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_named_ego_graph.html">make_named_ego_graph</a></span><span class="op">(</span><span class="va">gTree_1</span>, 
                              order <span class="op">=</span> <span class="va">neighbourhood_size</span>, 
                              min_ego_nodes <span class="op">=</span> <span class="va">min_ego_nodes</span>, 
                              min_ego_edges <span class="op">=</span> <span class="va">min_ego_edges</span><span class="op">)</span>

<span class="va">ego_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_named_ego_graph.html">make_named_ego_graph</a></span><span class="op">(</span><span class="va">gTree_2</span>, 
                              order <span class="op">=</span> <span class="va">neighbourhood_size</span>, 
                              min_ego_nodes <span class="op">=</span> <span class="va">min_ego_nodes</span>, 
                              min_ego_edges <span class="op">=</span> <span class="va">min_ego_edges</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">ego_1</span>,n<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; IGRAPH 1cfafa4 U--- 5 4 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from 1cfafa4:</span>
<span class="co">#&gt; [1] 1--2 2--3 2--4 2--5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; IGRAPH 9be9d13 U--- 5 4 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from 9be9d13:</span>
<span class="co">#&gt; [1] 1--2 2--3 2--4 2--5</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">ego_2</span>,n<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; IGRAPH 4d594c1 U--- 5 4 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from 4d594c1:</span>
<span class="co">#&gt; [1] 1--2 2--3 2--4 2--5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; IGRAPH dcec58e U--- 5 4 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from dcec58e:</span>
<span class="co">#&gt; [1] 1--2 2--3 2--4 2--5</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="count-the-number-of-nodes-and-the-subgraphs-in-ego-networks-of-query-graphs-n_w">Count the number of nodes and the subgraphs in ego-networks of query graphs (<span class="math inline">\(N_w\)</span>)<a class="anchor" aria-label="anchor" href="#count-the-number-of-nodes-and-the-subgraphs-in-ego-networks-of-query-graphs-n_w"></a>
</h3>
<p>Once the ego-networks are extracted, the subgraph counts for all ego-network are obtained for each network being compared:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Subgraphs counts for ego-networks in query graphs</span>
<span class="va">subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_to_graphlet_counts.html">ego_to_graphlet_counts</a></span><span class="op">(</span>ego_networks <span class="op">=</span> <span class="va">ego_1</span>, max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>
<span class="va">subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_to_graphlet_counts.html">ego_to_graphlet_counts</a></span><span class="op">(</span>ego_networks <span class="op">=</span> <span class="va">ego_2</span>, max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">subgraph_counts_1</span><span class="op">)</span>
<span class="co">#&gt;        N G0 G1 G2 G3 G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [395,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [396,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [397,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [398,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [399,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [400,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">subgraph_counts_2</span><span class="op">)</span>
<span class="co">#&gt;         N G0 G1 G2 G3 G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1595,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [1596,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [1597,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [1598,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [1599,] 5  4  6  0  0  4  0  0  0  0</span>
<span class="co">#&gt; [1600,] 5  4  6  0  0  4  0  0  0  0</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-expected-subgraph-counts">Calculate expected subgraph counts<a class="anchor" aria-label="anchor" href="#calculate-expected-subgraph-counts"></a>
</h3>
<p>In Netdis users can consider the following variants:</p>
<ul>
<li>Expectation via a gold-standard network (original proposal).</li>
<li>Expectation via the Geometric-Poisson / Polya-Aeppli approximation.</li>
<li>No expectation (equivalent to setting the expectation equal to zero).</li>
</ul>
<p>In this vignette the complete steps are given for “Expectation via a gold-standard network”. The step by step computation of Netdis-GP is in <a href="NetdisGPStepByStep.html">Netdis-GP: step by step</a>.</p>
<div class="section level4">
<h4 id="expectation-via-a-gold-standard-network">Expectation via a gold-standard network<a class="anchor" aria-label="anchor" href="#expectation-via-a-gold-standard-network"></a>
</h4>
<p>For this case the user must provide the gold-standard network of their choosing. This network will be used as a comparison reference point by Netdis.</p>
<p>The following considers a tree-like network with 900 nodes as the gold-standard.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Network used as gold-standard</span>
<span class="va">gst_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/erdos.renyi.game.html" class="external-link">erdos.renyi.game</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">30</span><span class="op">^</span><span class="fl">2</span>,p.or.m <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_density.html" class="external-link">graph.density</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gTree_2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gst_1</span>,vertex.size<span class="op">=</span><span class="fl">0.8</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="NetdisStepByStep_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="obtain-the-gold-standard-ego-network-counts-and-their-binning-according-to-their-edge-density-rho-">Obtain the gold-standard ego-network counts and their binning according to their edge-density (<span class="math inline">\(\rho(.)\)</span>)<a class="anchor" aria-label="anchor" href="#obtain-the-gold-standard-ego-network-counts-and-their-binning-according-to-their-edge-density-rho-"></a>
</h4>
<p>To calculate the expected counts, <span class="math inline">\(E_w\)</span>, the counts of the ego-networks of the gold-standard network need to be obtained first:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Obtain subgraph counts and binning for gold-standard</span>
<span class="va">ego_gst_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_named_ego_graph.html">make_named_ego_graph</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gst_1</span>, 
                              order <span class="op">=</span> <span class="va">neighbourhood_size</span>, 
                              min_ego_nodes <span class="op">=</span> <span class="va">min_ego_nodes</span>, 
                              min_ego_edges <span class="op">=</span> <span class="va">min_ego_edges</span><span class="op">)</span>

<span class="va">subgraph_counts_gst_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_to_graphlet_counts.html">ego_to_graphlet_counts</a></span><span class="op">(</span>ego_networks <span class="op">=</span> <span class="va">ego_gst_1</span>,
                                                max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">subgraph_counts_gst_1</span><span class="op">)</span>
<span class="co">#&gt;       N G0 G1 G2 G3 G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1,]  5  4  3  0  2  0  0  0  0  0</span>
<span class="co">#&gt; [2,]  8  7 10  0  5  5  0  0  0  0</span>
<span class="co">#&gt; [3,]  3  2  1  0  0  0  0  0  0  0</span>
<span class="co">#&gt; [4,] 11 10 15  0 18  8  0  0  0  0</span>
<span class="co">#&gt; [5,]  3  2  1  0  0  0  0  0  0  0</span>
<span class="co">#&gt; [6,]  3  2  1  0  0  0  0  0  0  0</span></code></pre></div>
<p>Subsequently, these ego-networks are binned according to their edge density:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">densities_gst_1</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_network_density.html">ego_network_density</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_gst_1</span><span class="op">)</span>

<span class="co"># Adaptively bin ego-network densities</span>
<span class="va">binned_densities_gst_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/binned_densities_adaptive.html">binned_densities_adaptive</a></span><span class="op">(</span>densities <span class="op">=</span> <span class="va">densities_gst_1</span>, 
                                                min_counts_per_interval <span class="op">=</span> <span class="va">min_bin_count</span>, 
                                                num_intervals <span class="op">=</span> <span class="va">num_bins</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">binned_densities_gst_1</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ densities       : num [1:516] 0.4 0.25 0.667 0.182 0.667 ...</span>
<span class="co">#&gt;  $ interval_indexes: int [1:516] 7 4 9 1 9 9 9 9 8 6 ...</span>
<span class="co">#&gt;  $ breaks          : num [1:10] 0.167 0.187 0.202 0.227 0.252 ...</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="obtain-the-scaled-mean-counts-for-each-density-bin-for-the-gold-standard-network-canonical-e_w">Obtain the scaled mean counts for each density bin for the gold-standard network (canonical <span class="math inline">\(E_w\)</span>)<a class="anchor" aria-label="anchor" href="#obtain-the-scaled-mean-counts-for-each-density-bin-for-the-gold-standard-network-canonical-e_w"></a>
</h4>
<p><span class="math inline">\(E_w\)</span> is estimated based on the average subgraph counts of ego-networks per density bin for each given subgraph. However, as the query networks and the gold-standard may have different number of nodes, the counts of the gold-standard network are first scaled to a “standard” or “canonical” scale from which they can be scaled back towards networks of different sizes. The following code first shows the computation of the subgraph counts for the ego-networks in the gold-standard network with their corresponding scaling:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scale ego-network subgraph counts by dividing by the total number of k-tuples in the</span>
<span class="co"># ego-network (where k is the subgraph size)</span>
<span class="va">scaled_subgraph_counts_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scale_graphlet_counts_ego.html">scale_graphlet_counts_ego</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_gst_1</span>,
                                                        max_graphlet_size <span class="op">=</span><span class="va">max_subgraph_size</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">scaled_subgraph_counts_ref</span><span class="op">)</span>
<span class="co">#&gt;  num [1:516, 1:9] 0.4 0.25 0.667 0.182 0.667 ...</span>
<span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   ..$ : NULL</span>
<span class="co">#&gt;   ..$ : chr [1:9] "G0" "G1" "G2" "G3" ...</span></code></pre></div>
<p>Finally, the standard or canonical <span class="math inline">\(E_w\)</span> can be obtained by taking the average per bin of the scaled subgraph counts:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Average of the scaled reference subgraph counts in each density bin</span>
<span class="va">ref_binned_canonical_subgraph_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mean_density_binned_graphlet_counts.html">mean_density_binned_graphlet_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">scaled_subgraph_counts_ref</span>, density_interval_indexes <span class="op">=</span> <span class="va">binned_densities_gst_1</span><span class="op">$</span><span class="va">interval_indexes</span><span class="op">)</span>

<span class="va">ref_binned_canonical_subgraph_counts</span>
<span class="co">#&gt;          G0         G1           G2         G3         G4 G5           G6 G7 G8</span>
<span class="co">#&gt; 1 0.1774892 0.09393939 0.0000000000 0.05108225 0.02871573  0 0.0000000000  0  0</span>
<span class="co">#&gt; 2 0.2000000 0.10972222 0.0000000000 0.06666667 0.03095238  0 0.0000000000  0  0</span>
<span class="co">#&gt; 3 0.2222222 0.14732143 0.0000000000 0.07738095 0.06001984  0 0.0000000000  0  0</span>
<span class="co">#&gt; 4 0.2500000 0.17028061 0.0000000000 0.10204082 0.06224490  0 0.0000000000  0  0</span>
<span class="co">#&gt; 5 0.2857143 0.21879699 0.0000000000 0.13984962 0.09624060  0 0.0000000000  0  0</span>
<span class="co">#&gt; 6 0.3333333 0.31393443 0.0000000000 0.18579235 0.20655738  0 0.0000000000  0  0</span>
<span class="co">#&gt; 7 0.4000000 0.44382022 0.0005617978 0.25842697 0.35955056  0 0.0007490637  0  0</span>
<span class="co">#&gt; 8 0.5000000 0.67317073 0.0008130081 0.29593496 0.69918699  0 0.0016260163  0  0</span>
<span class="co">#&gt; 9 0.6666667 0.99324324 0.0033783784 0.00000000 0.00000000  0 0.0135135135  0  0</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="obtain-the-centred-counts-for-each-subgraph-on-the-query-graphs-based-on-the-gold-standard-counts-n_wg---e_wg">Obtain the centred counts for each subgraph on the query graphs based on the gold-standard counts (<span class="math inline">\(N_w(G) - E_w(G)\)</span>)<a class="anchor" aria-label="anchor" href="#obtain-the-centred-counts-for-each-subgraph-on-the-query-graphs-based-on-the-gold-standard-counts-n_wg---e_wg"></a>
</h3>
<p>After obtaining the average scaled subgraph counts per density bin, the subgraph counts of the query networks can be centred:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scale the reference counts of the gold-standard network to the sizes of each of the query ego-networks.</span>
<span class="va">exp_subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_expected_counts.html">netdis_expected_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_1</span>,
                                                density_breaks <span class="op">=</span> <span class="va">binned_densities_gst_1</span><span class="op">$</span><span class="va">breaks</span>, 
                                                density_binned_reference_counts <span class="op">=</span>  <span class="va">ref_binned_canonical_subgraph_counts</span>, 
                                                max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span>,
                                                scale_fn<span class="op">=</span><span class="va">count_graphlet_tuples</span><span class="op">)</span>


<span class="va">exp_subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_expected_counts.html">netdis_expected_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_2</span>,
                                                density_breaks <span class="op">=</span> <span class="va">binned_densities_gst_1</span><span class="op">$</span><span class="va">breaks</span>, 
                                                density_binned_reference_counts <span class="op">=</span>  <span class="va">ref_binned_canonical_subgraph_counts</span>, 
                                                max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span>,
                                                scale_fn<span class="op">=</span><span class="va">count_graphlet_tuples</span><span class="op">)</span>

<span class="co"># Centre subgraph counts by subtracting expected counts</span>
<span class="va">centred_subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_subtract_exp_counts.html">netdis_subtract_exp_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_1</span>,
                                                        exp_graphlet_counts <span class="op">=</span> <span class="va">exp_subgraph_counts_1</span>, 
                                                        max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="va">centred_subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_subtract_exp_counts.html">netdis_subtract_exp_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_2</span>,
                                                        exp_graphlet_counts <span class="op">=</span> <span class="va">exp_subgraph_counts_2</span>, 
                                                        max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_1</span><span class="op">)</span>
<span class="co">#&gt;        G0       G1           G2        G3       G4 G5           G6 G7 G8</span>
<span class="co">#&gt; [395,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [396,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [397,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [398,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [399,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [400,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_2</span><span class="op">)</span>
<span class="co">#&gt;         G0       G1           G2        G3       G4 G5           G6 G7 G8</span>
<span class="co">#&gt; [1595,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [1596,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [1597,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [1598,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [1599,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span>
<span class="co">#&gt; [1600,]  0 1.561798 -0.005617978 -1.292135 2.202247  0 -0.003745318  0  0</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sum-centred-subgraph-counts-across-all-ego-networks-s_wg">Sum centred subgraph counts across all ego-networks (<span class="math inline">\(S_w(G)\)</span>)<a class="anchor" aria-label="anchor" href="#sum-centred-subgraph-counts-across-all-ego-networks-s_wg"></a>
</h3>
<p>After the counts are centred, the total count is computed for each subgraph in each query network:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sum_subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_1</span><span class="op">)</span>
<span class="va">sum_subgraph_counts_1</span>
<span class="co">#&gt;         G0         G1         G2         G3         G4         G5         G6 </span>
<span class="co">#&gt;  -346.2814  -801.5149    -1.5000 -3801.1775 -1134.8427     0.0000    -1.0000 </span>
<span class="co">#&gt;         G7         G8 </span>
<span class="co">#&gt;     0.0000     0.0000</span>

<span class="va">sum_subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_2</span><span class="op">)</span>
<span class="va">sum_subgraph_counts_2</span>
<span class="co">#&gt;            G0            G1            G2            G3            G4 </span>
<span class="co">#&gt;  -1428.467532  -3399.955521     -5.994382 -15721.318255  -4831.197918 </span>
<span class="co">#&gt;            G5            G6            G7            G8 </span>
<span class="co">#&gt;      0.000000     -3.996255      0.000000      0.000000</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-the-netdis-statistics">Calculate the Netdis statistics<a class="anchor" aria-label="anchor" href="#calculate-the-netdis-statistics"></a>
</h3>
<p>Finally, the total centred counts can be used to obtain the Netdis statistic:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">netdis_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_uptok.html">netdis_uptok</a></span><span class="op">(</span>centred_graphlet_count_vector_1 <span class="op">=</span> <span class="va">sum_subgraph_counts_1</span>,
                              centred_graphlet_count_vector_2 <span class="op">=</span> <span class="va">sum_subgraph_counts_2</span>, 
                              max_graphlet_size <span class="op">=</span>  <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">netdis_result</span><span class="op">)</span>
<span class="co">#&gt;      netdis3      netdis4 </span>
<span class="co">#&gt; 1.653858e-06 3.705944e-05</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<ul>
<li><p>W. Ali, T. Rito, G. Reinert, F. Sun, and C. M. Deane. Alignment-free protein interaction network comparison. Bioinformatics, 30:i430–i437, 2014.</p></li>
<li><p>L. Ospina-Forero, C. M. Deane, and G. Reinert. Assessment of model fit via network comparison methods based on subgraph counts. Journal of Complex Networks, page cny017, August 2018.</p></li>
<li><p>A. E. Wegner, L. Ospina-Forero, R. E. Gaunt, C. M. Deane, and G. Reinert. Identifying networks with common organizational principles. Journal of Complex networks, 2017.</p></li>
<li><p>F. Picard, J.-J. Daudin, M. Koskas, S. Schbath, and S. Robin. Assessing the exceptionality of network motifs. Journal of Computational Biology, 15(1):1–20, 2008.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin O’Reilly.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
