<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Netdis-GP: step by step • netdist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Netdis-GP: step by step">
<meta property="og:description" content="netdist">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">netdist</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ManyToMany.html">Simple and quick (default) usage 3: many to many comparisons</a>
    </li>
    <li>
      <a href="../articles/NetEmdTimeOrdering.html">NetEmd: World trade</a>
    </li>
    <li>
      <a href="../articles/NetdisGPStepByStep.html">Netdis-GP: step by step</a>
    </li>
    <li>
      <a href="../articles/NetdisStepByStep.html">Netdis: step by step</a>
    </li>
    <li>
      <a href="../articles/NewNetdisCustomisations.html">Netdis: customizations</a>
    </li>
    <li>
      <a href="../articles/PreComputedProps.html">Simple and quick (default) usage 2: pre-computed properties</a>
    </li>
    <li>
      <a href="../articles/V-Menu.html">Netdis Vignette's Menu</a>
    </li>
    <li>
      <a href="../articles/default_pairwise_usage.html">Simple and quick (default) usage 1: pairwise comparisons</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_dis.html">Netdis: Protein interaction networks</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_emd.html">Dendrogram example for NetEMD</a>
    </li>
    <li>
      <a href="../articles/netdis_customisations.html">Usage of netdis with binning and expected counts customisations.</a>
    </li>
    <li>
      <a href="../articles/netdis_pairwise_comparisons.html">Usage of netdis interfaces for different pairwise comparison options.</a>
    </li>
    <li>
      <a href="../articles/quickstart_netdis_2graphs.html">Quick start guide for Netdis - 2 graphs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Netdis-GP: step by step</h1>
            
            <h4 data-toc-skip class="date">2022-06-06</h4>
      
      
      <div class="hidden name"><code>NetdisGPStepByStep.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette explains some inner calculations of Netdis-GP, which the user does not need to go through, however we explain them here for those users that may want to use Netdis in a modular fashion. Netdis-GP is a version of Netdis that uses the Polya-Aeppli/Geometric-Poisson approximation for the subgraph/graphlet count expectations.</p>
<p>For a simple Netdis-GP function call see instead <a href="default_pairwise_usage.html">“Simple and quick (default) usage 1: pairwise comparisons”</a>.</p>
<p>For other vignettes in this package see the <a href="V-Menu.html">“Menu”</a>.</p>
</div>
<div class="section level2">
<h2 id="what-is-netdis-gp">What is Netdis-GP?<a class="anchor" aria-label="anchor" href="#what-is-netdis-gp"></a>
</h2>
<p>(Extracted from Ali et al. (2014) and Ospina-Forero et al. (2018)): Netdis counts small subgraphs <span class="math inline">\(w\)</span> on <span class="math inline">\(k\)</span> nodes for all 2-step ego-networks, <span class="math inline">\(k=3,4,5\)</span>. These counts are centred by subtracting the expected number of counts <span class="math inline">\(E_w\)</span>. These centred counts of each network are then compared thus leading to the Netdis statistic.</p>
<p>In Netdis-GP, the expected number of counts <span class="math inline">\(E_w\)</span> is computed independently for each graph based on a Geometric-Poisson (GP) approximation for the distribution of the number of occurrences of subgraph <span class="math inline">\(w\)</span> in the respective graph. It assumes that <span class="math inline">\(N_{w,i} \sim GP(\lambda^{\rho(i)}_k, \theta^{\rho(i)}_w)\)</span>, where <span class="math inline">\(\lambda^{\rho(i)}_k\)</span> is the Poisson parameter indexed by the size of subgraph <span class="math inline">\(w\)</span> and the density bin <span class="math inline">\(\rho(i)\)</span>; and where <span class="math inline">\(\theta^{\rho(i)}_w\)</span> is the geometric parameter indexed by subgraph <span class="math inline">\(w\)</span> and density bin <span class="math inline">\(\rho(i)\)</span>. <span class="math inline">\(E_w(G, \rho(i))\)</span> is taken as the mean of the GP approximation, i.e. <span class="math inline">\(\lambda^{\rho(i)}_k/\theta^{\rho(i)}_w\)</span>. As <span class="math inline">\(\lambda^{\rho(i)}_k\)</span> and <span class="math inline">\(\theta^{\rho(i)}_w\)</span> are not known, they are estimated as follows:</p>
<p>Let <span class="math inline">\(x_{w,d}^j\)</span> be the number of subgraphs <span class="math inline">\(w\)</span> on the 2-step ego-network <span class="math inline">\(j\)</span> of density bin <span class="math inline">\(d\)</span>, and let <span class="math display">\[
\bar{X}_{w,d}=\frac{1}{q} \sum_{j=1}^q x_{w,d}^j, \qquad V^2_{w,d}=\frac{1}{q-1} \sum_{j=1}^q (x_{w,d}^j - \bar{X}_{w,d})^2
,
\]</span> where <span class="math inline">\(q\)</span> is the number of ego-networks in density bin <span class="math inline">\(d\)</span>. Then, <span class="math display">\[
\hat{\lambda}^{d}_{k}= \frac{1}{l} \sum_{h \in A(k)} \frac{2 (\bar{X}_{h,d})^2}{V^2_{h,d}+\bar{X}_{h,d}} , \qquad \hat{\theta}^{d}_w= \frac{2\bar{X}_{w,d}}{V^2_{w,d}+\bar{X}_{w,d}},
\]</span> where <span class="math inline">\(l\)</span> is the number of connected subgraphs of size <span class="math inline">\(k\)</span>, for example, <span class="math inline">\(l=6\)</span> for <span class="math inline">\(k=4\)</span>. These estimators are based on the moment estimators of a GP random variable and the proposal made by Picard et al.(2008), where the total count of each individual subgraph could be thought as the sum of the total subgraph counts over multiple ``clumps’’ of edges that appear across the network.</p>
<p>This variant focuses on detecting more general and global discrepancies between the ego-network structures.</p>
<div class="section level3">
<h3 id="netdis-and-its-variants-are-constructed-as-follows">Netdis and its variants are constructed as follows:<a class="anchor" aria-label="anchor" href="#netdis-and-its-variants-are-constructed-as-follows"></a>
</h3>
<p>(Extracted from Ali et al. (2014)): Let <span class="math inline">\(N_{w,i}(G)\)</span> be the number of induced occurrences of small graphs <span class="math inline">\(w\)</span> in the 2-step ego network of vertex <span class="math inline">\(i\)</span>. Now, bin all 2-step ego-networks of network <span class="math inline">\(G\)</span> according to their network density. Let <span class="math inline">\(E_w(G,\rho)\)</span> be the expected number of occurrences of <span class="math inline">\(w\)</span> in an ego-network whose density falls in density bin <span class="math inline">\(\rho\)</span>. For a given network <span class="math inline">\(G\)</span> compute the centred subgraph counts as <span class="math display">\[
S_w(G)=\sum\limits_{i }{\bigg (N_{w,i}(G)- E_w(G, \rho(i)) \bigg )},
\]</span> where <span class="math inline">\(i\)</span> is a node in <span class="math inline">\(G\)</span> and <span class="math inline">\(\rho(i)\)</span> the density bin of the 2-step ego-network of node <span class="math inline">\(i\)</span>.</p>
<p>Now, to compare networks <span class="math inline">\(G_1\)</span> and <span class="math inline">\(G_2\)</span>, set <span class="math display">\[
\displaystyle
netD_2^S(k) = \tfrac{1}{ \sqrt{ M(k)} } \sum\limits_{w \in A(k)}
\bigg ({ \tfrac{S_w(G_1)  S_w(G_2)} {\sqrt{S_w(G_1)^2 + S_w(G_2)^2}}
}\bigg ), \quad k=3,4, 5, 
\]</span> where <span class="math inline">\(A(k)\)</span> is the set of connected subgraphs of size <span class="math inline">\(k\)</span>, and where <span class="math inline">\(M(k)\)</span> is a normalising constant so that <span class="math inline">\(netD_2^S(k)\in[-1,1]\)</span>. <span class="math inline">\(M(k)\)</span> is equal to <span class="math display">\[
M(k)  = \sum\limits_{w \in A(k)}
\left( \tfrac{ S_w(G_1)^2 }{\sqrt{S_w(G_1)^2 + S_w(G_2)^2}} \right)
\sum\limits_{w \in A(k)}
\left(\tfrac{ S_w(G_2)^2 } {\sqrt{S_w(G_1)^2 + S_w(G_2)^2}}
\right)
.
\]</span> The corresponding Netdis statistic is defined as <span class="math display">\[Netdis(k)=netd_2^S(k)=\tfrac{1}{2}(1-netD_2^S(k)) \in [0,1].\]</span> Small values of Netdis suggest higher `similarity’ between the networks. By default Netdis uses subgraphs on <span class="math inline">\(k=4\)</span> nodes.</p>
</div>
</div>
<div class="section level2">
<h2 id="load-required-libraries">Load required libraries<a class="anchor" aria-label="anchor" href="#load-required-libraries"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"netdist"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://igraph.org" class="external-link">"igraph"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://purrr.tidyverse.org" class="external-link">"purrr"</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="netdis-gp-step-by-step">Netdis-GP step by step<a class="anchor" aria-label="anchor" href="#netdis-gp-step-by-step"></a>
</h2>
<div class="section level3">
<h3 id="networks-being-compared">Networks being compared<a class="anchor" aria-label="anchor" href="#networks-being-compared"></a>
</h3>
<p>Generation of tree-like networks with 400 nodes and 1600 nodes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create networks</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">34</span><span class="op">)</span>
<span class="va">gTree_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/as.directed.html" class="external-link">as.undirected</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_tree.html" class="external-link">make_tree</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">20</span><span class="op">^</span><span class="fl">2</span>,children <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">)</span>
<span class="va">gTree_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/as.directed.html" class="external-link">as.undirected</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_tree.html" class="external-link">make_tree</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">40</span><span class="op">^</span><span class="fl">2</span>,children <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gTree_1</span>,vertex.size<span class="op">=</span><span class="fl">0.8</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="NetdisGPStepByStep_files/figure-html/netwokrs-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gTree_2</span>,vertex.size<span class="op">=</span><span class="fl">0.8</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="NetdisGPStepByStep_files/figure-html/netwokrs-2.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="set-netdis-parameters">Set Netdis parameters<a class="anchor" aria-label="anchor" href="#set-netdis-parameters"></a>
</h3>
<p>Netdis uses some mostly internal parameters that define the size of the subgraphs/graphlets that are going to be used, as well as the length of the ego neighbourhood, the minimum size considered for the resulting ego-networks, and finally some parameters that control an ego-network binning function. By default, Netdis considers the following setup:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Maximum subgraph size to calculate counts and netdis statistic for</span>
<span class="va">max_subgraph_size</span> <span class="op">&lt;-</span> <span class="fl">4</span>

<span class="co"># Ego-network neighbourhood size</span>
<span class="va">neighbourhood_size</span> <span class="op">&lt;-</span> <span class="fl">2</span>

<span class="co"># Minimum size of ego-networks to consider</span>
<span class="va">min_ego_nodes</span> <span class="op">&lt;-</span> <span class="fl">3</span>
<span class="va">min_ego_edges</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="co"># Ego-network density binning parameters. Here, the minimum number of ego-networks per bin and the starting number of bins</span>
<span class="va">min_bin_count</span> <span class="op">&lt;-</span> <span class="fl">5</span>
<span class="va">num_bins</span> <span class="op">&lt;-</span> <span class="fl">100</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="obtain-list-of-ego-networks">Obtain list of ego networks<a class="anchor" aria-label="anchor" href="#obtain-list-of-ego-networks"></a>
</h3>
<p>One of the first steps in Netdis is the extraction of all ego-networks in each of the query networks:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get ego-networks for query graphs</span>
<span class="va">ego_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_named_ego_graph.html">make_named_ego_graph</a></span><span class="op">(</span><span class="va">gTree_1</span>, 
                              order <span class="op">=</span> <span class="va">neighbourhood_size</span>, 
                              min_ego_nodes <span class="op">=</span> <span class="va">min_ego_nodes</span>, 
                              min_ego_edges <span class="op">=</span> <span class="va">min_ego_edges</span><span class="op">)</span>

<span class="va">ego_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_named_ego_graph.html">make_named_ego_graph</a></span><span class="op">(</span><span class="va">gTree_2</span>, 
                              order <span class="op">=</span> <span class="va">neighbourhood_size</span>, 
                              min_ego_nodes <span class="op">=</span> <span class="va">min_ego_nodes</span>, 
                              min_ego_edges <span class="op">=</span> <span class="va">min_ego_edges</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ego_1</span>,n<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; IGRAPH 1df2706 U--- 13 12 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from 1df2706:</span>
<span class="co">#&gt;  [1] 1-- 2 1-- 3 1-- 4 2-- 5 2-- 6 2-- 7 3-- 8 3-- 9 3--10 4--11 4--12 4--13</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; IGRAPH 61f1ce7 U--- 16 15 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from 61f1ce7:</span>
<span class="co">#&gt;  [1] 1-- 2 1-- 3 1-- 4 2-- 5 2-- 6 2-- 7 5-- 8 5-- 9 5--10 6--11 6--12 6--13</span>
<span class="co">#&gt; [13] 7--14 7--15 7--16</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ego_2</span>,n<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; IGRAPH e1910b3 U--- 13 12 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from e1910b3:</span>
<span class="co">#&gt;  [1] 1-- 2 1-- 3 1-- 4 2-- 5 2-- 6 2-- 7 3-- 8 3-- 9 3--10 4--11 4--12 4--13</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; IGRAPH 8e40848 U--- 16 15 -- Tree</span>
<span class="co">#&gt; + attr: name (g/c), children (g/n), mode (g/c)</span>
<span class="co">#&gt; + edges from 8e40848:</span>
<span class="co">#&gt;  [1] 1-- 2 1-- 3 1-- 4 2-- 5 2-- 6 2-- 7 5-- 8 5-- 9 5--10 6--11 6--12 6--13</span>
<span class="co">#&gt; [13] 7--14 7--15 7--16</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="count-the-number-of-nodes-and-the-subgraphs-in-the-ego-networks-of-each-graph-n_w">Count the number of nodes and the subgraphs in the ego-networks of each graph (<span class="math inline">\(N_w\)</span>)<a class="anchor" aria-label="anchor" href="#count-the-number-of-nodes-and-the-subgraphs-in-the-ego-networks-of-each-graph-n_w"></a>
</h3>
<p>Once the ego-networks are extracted, the subgraph counts for all ego-network are obtained for each network being compared:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Subgraphs counts for ego-networks in query graphs</span>
<span class="va">subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_to_graphlet_counts.html">ego_to_graphlet_counts</a></span><span class="op">(</span>ego_networks <span class="op">=</span> <span class="va">ego_1</span>, max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>
<span class="va">subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_to_graphlet_counts.html">ego_to_graphlet_counts</a></span><span class="op">(</span>ego_networks <span class="op">=</span> <span class="va">ego_2</span>, max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">subgraph_counts_1</span><span class="op">)</span>
<span class="co">#&gt;       N G0 G1 G2 G3 G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1,] 13 12 21  0 18 13  0  0  0  0</span>
<span class="co">#&gt; [2,] 16 15 27  0 33 17  0  0  0  0</span>
<span class="co">#&gt; [3,] 16 15 27  0 33 17  0  0  0  0</span>
<span class="co">#&gt; [4,] 16 15 27  0 33 17  0  0  0  0</span>
<span class="co">#&gt; [5,] 17 16 30  0 36 20  0  0  0  0</span>
<span class="co">#&gt; [6,] 17 16 30  0 36 20  0  0  0  0</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">subgraph_counts_2</span><span class="op">)</span>
<span class="co">#&gt;       N G0 G1 G2 G3 G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1,] 13 12 21  0 18 13  0  0  0  0</span>
<span class="co">#&gt; [2,] 16 15 27  0 33 17  0  0  0  0</span>
<span class="co">#&gt; [3,] 16 15 27  0 33 17  0  0  0  0</span>
<span class="co">#&gt; [4,] 16 15 27  0 33 17  0  0  0  0</span>
<span class="co">#&gt; [5,] 17 16 30  0 36 20  0  0  0  0</span>
<span class="co">#&gt; [6,] 17 16 30  0 36 20  0  0  0  0</span></code></pre></div>
<div class="section level4">
<h4 id="obtain-binning-for-each-ego-network-according-to-the-edge-density-of-the-ego-networks-of-each-individual-graph-rho-">Obtain binning for each ego-network according to the edge-density of the ego-networks of each individual graph (<span class="math inline">\(\rho(.)\)</span>)<a class="anchor" aria-label="anchor" href="#obtain-binning-for-each-ego-network-according-to-the-edge-density-of-the-ego-networks-of-each-individual-graph-rho-"></a>
</h4>
<p>For each ego-network, its edge density is obtained, then the ego-networks are binned according to their edge density. This process is repeated for each of the query networks:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get ego-network densities</span>
<span class="va">densities_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_network_density.html">ego_network_density</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_1</span><span class="op">)</span>
<span class="va">densities_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ego_network_density.html">ego_network_density</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_2</span><span class="op">)</span>

<span class="co"># Adaptively bin ego-network densities</span>
<span class="va">binned_densities_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/binned_densities_adaptive.html">binned_densities_adaptive</a></span><span class="op">(</span>densities <span class="op">=</span> <span class="va">densities_1</span>, 
                                                min_counts_per_interval <span class="op">=</span> <span class="va">min_bin_count</span>, 
                                                num_intervals <span class="op">=</span> <span class="va">num_bins</span><span class="op">)</span>

<span class="va">binned_densities_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/binned_densities_adaptive.html">binned_densities_adaptive</a></span><span class="op">(</span>densities <span class="op">=</span> <span class="va">densities_2</span>, 
                                                min_counts_per_interval <span class="op">=</span> <span class="va">min_bin_count</span>, 
                                                num_intervals <span class="op">=</span> <span class="va">num_bins</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">binned_densities_1</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ densities       : num [1:400] 0.154 0.125 0.125 0.125 0.118 ...</span>
<span class="co">#&gt;  $ interval_indexes: int [1:400] 2 2 2 2 1 1 1 1 1 1 ...</span>
<span class="co">#&gt;  $ breaks          : num [1:4] 0.118 0.12 0.25 0.4</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">binned_densities_2</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ densities       : num [1:1600] 0.154 0.125 0.125 0.125 0.118 ...</span>
<span class="co">#&gt;  $ interval_indexes: int [1:1600] 2 2 2 2 1 1 1 1 1 1 ...</span>
<span class="co">#&gt;  $ breaks          : num [1:5] 0.118 0.12 0.183 0.25 0.4</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="calculate-expected-subgraph-counts-in-each-density-bin-by-the-geometric-poisson-approximation-e_w">Calculate expected subgraph counts in each density bin by the Geometric-Poisson approximation (<span class="math inline">\(E_w\)</span>)<a class="anchor" aria-label="anchor" href="#calculate-expected-subgraph-counts-in-each-density-bin-by-the-geometric-poisson-approximation-e_w"></a>
</h3>
<p>With the ego-network binning obtained, the Geometric-Poisson approximation of the expected subgraph counts, <span class="math inline">\(E_w\)</span>, can be obtained for each subgraph <span class="math inline">\(w\)</span> and each density bin:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">binned_gp_subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/density_binned_counts_gp.html">density_binned_counts_gp</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_1</span>, 
                                                     density_interval_indexes <span class="op">=</span>  <span class="va">binned_densities_1</span><span class="op">$</span><span class="va">interval_indexes</span>, 
                                                     max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="va">binned_gp_subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/density_binned_counts_gp.html">density_binned_counts_gp</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_2</span>, 
                                                     density_interval_indexes <span class="op">=</span>  <span class="va">binned_densities_2</span><span class="op">$</span><span class="va">interval_indexes</span>, 
                                                     max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>
<span class="va">binned_gp_subgraph_counts_1</span>
<span class="co">#&gt;             G0        G1 G2       G3        G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1,] 16.000000 15.000000  0 9.325963 9.3466297  0  0  0  0</span>
<span class="co">#&gt; [2,]  7.311828  6.290323  0 4.681556 2.1441715  0  0  0  0</span>
<span class="co">#&gt; [3,]  4.000000  3.000000  0 0.000000 0.6666667  0  0  0  0</span>
<span class="va">binned_gp_subgraph_counts_2</span>
<span class="co">#&gt;        G0 G1 G2        G3        G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1,] 16.0 15  0  9.325963 9.3466297  0  0  0  0</span>
<span class="co">#&gt; [2,] 13.4 12  0 10.797819 4.3434837  0  0  0  0</span>
<span class="co">#&gt; [3,]  7.0  6  0  2.831504 2.8353938  0  0  0  0</span>
<span class="co">#&gt; [4,]  4.0  3  0  0.000000 0.6666667  0  0  0  0</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="obtain-the-centred-counts-for-each-subgraph-on-the-query-graphs-based-on-the-geometric-poisson-approximation-n_wg---e_wg">Obtain the centred counts for each subgraph on the query graphs based on the Geometric-Poisson approximation (<span class="math inline">\(N_w(G) - E_w(G)\)</span>)<a class="anchor" aria-label="anchor" href="#obtain-the-centred-counts-for-each-subgraph-on-the-query-graphs-based-on-the-geometric-poisson-approximation-n_wg---e_wg"></a>
</h3>
<p>With <span class="math inline">\(E_w\)</span> now obtained, Netdis-GP, can be compute as per its construction by first centring the observed counts:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Calculate expected subgraph counts for each ego network</span>
<span class="va">exp_gp_subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_expected_counts.html">netdis_expected_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_1</span>,
                                                density_breaks <span class="op">=</span> <span class="va">binned_densities_1</span><span class="op">$</span><span class="va">breaks</span>, 
                                                density_binned_reference_counts <span class="op">=</span>  <span class="va">binned_gp_subgraph_counts_1</span>, 
                                                max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span>,
                                                scale_fn<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>


<span class="va">exp_gp_subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_expected_counts.html">netdis_expected_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_2</span>,
                                                density_breaks <span class="op">=</span> <span class="va">binned_densities_2</span><span class="op">$</span><span class="va">breaks</span>, 
                                                density_binned_reference_counts <span class="op">=</span>  <span class="va">binned_gp_subgraph_counts_2</span>, 
                                                max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span>,
                                                scale_fn<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>

<span class="co"># Centre subgraph counts by subtracting expected counts</span>
<span class="va">centred_subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_subtract_exp_counts.html">netdis_subtract_exp_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_1</span>,
                                                        exp_graphlet_counts <span class="op">=</span> <span class="va">exp_gp_subgraph_counts_1</span>, 
                                                        max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="va">centred_subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_subtract_exp_counts.html">netdis_subtract_exp_counts</a></span><span class="op">(</span>graphlet_counts <span class="op">=</span> <span class="va">subgraph_counts_2</span>,
                                                        exp_graphlet_counts <span class="op">=</span> <span class="va">exp_gp_subgraph_counts_2</span>, 
                                                        max_graphlet_size <span class="op">=</span> <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_1</span><span class="op">)</span>
<span class="co">#&gt;            G0       G1 G2       G3       G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1,] 4.688172 14.70968  0 13.31844 10.85583  0  0  0  0</span>
<span class="co">#&gt; [2,] 7.688172 20.70968  0 28.31844 14.85583  0  0  0  0</span>
<span class="co">#&gt; [3,] 7.688172 20.70968  0 28.31844 14.85583  0  0  0  0</span>
<span class="co">#&gt; [4,] 7.688172 20.70968  0 28.31844 14.85583  0  0  0  0</span>
<span class="co">#&gt; [5,] 0.000000 15.00000  0 26.67404 10.65337  0  0  0  0</span>
<span class="co">#&gt; [6,] 0.000000 15.00000  0 26.67404 10.65337  0  0  0  0</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_2</span><span class="op">)</span>
<span class="co">#&gt;        G0 G1 G2        G3        G4 G5 G6 G7 G8</span>
<span class="co">#&gt; [1,] -1.4  9  0  7.202181  8.656516  0  0  0  0</span>
<span class="co">#&gt; [2,]  1.6 15  0 22.202181 12.656516  0  0  0  0</span>
<span class="co">#&gt; [3,]  1.6 15  0 22.202181 12.656516  0  0  0  0</span>
<span class="co">#&gt; [4,]  1.6 15  0 22.202181 12.656516  0  0  0  0</span>
<span class="co">#&gt; [5,]  0.0 15  0 26.674037 10.653370  0  0  0  0</span>
<span class="co">#&gt; [6,]  0.0 15  0 26.674037 10.653370  0  0  0  0</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sum-centred-subgraph-counts-across-all-ego-networks-s_wg">Sum centred subgraph counts across all ego networks (<span class="math inline">\(S_w(G)\)</span>)<a class="anchor" aria-label="anchor" href="#sum-centred-subgraph-counts-across-all-ego-networks-s_wg"></a>
</h3>
<p>After the counts are centred, the total count is computed for each subgraph in each query network:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sum_subgraph_counts_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_1</span><span class="op">)</span>
<span class="va">sum_subgraph_counts_1</span>
<span class="co">#&gt;           G0           G1           G2           G3           G4           G5 </span>
<span class="co">#&gt; 2.575717e-14 1.986000e+03 0.000000e+00 1.549577e+03 1.892727e+03 0.000000e+00 </span>
<span class="co">#&gt;           G6           G7           G8 </span>
<span class="co">#&gt; 0.000000e+00 0.000000e+00 0.000000e+00</span>

<span class="va">sum_subgraph_counts_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">centred_subgraph_counts_2</span><span class="op">)</span>
<span class="va">sum_subgraph_counts_2</span>
<span class="co">#&gt;            G0            G1            G2            G3            G4 </span>
<span class="co">#&gt; -3.170797e-13  7.986000e+03  0.000000e+00  6.885435e+03  7.287418e+03 </span>
<span class="co">#&gt;            G5            G6            G7            G8 </span>
<span class="co">#&gt;  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-netdis-gp-statistics">Calculate Netdis-GP statistics<a class="anchor" aria-label="anchor" href="#calculate-netdis-gp-statistics"></a>
</h3>
<p>Finally, the total centred counts can be used to obtain the Netdis statistic:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">netdis_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netdis_uptok.html">netdis_uptok</a></span><span class="op">(</span>centred_graphlet_count_vector_1 <span class="op">=</span> <span class="va">sum_subgraph_counts_1</span>,
                              centred_graphlet_count_vector_2 <span class="op">=</span> <span class="va">sum_subgraph_counts_2</span>, 
                              max_graphlet_size <span class="op">=</span>  <span class="va">max_subgraph_size</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">netdis_result</span><span class="op">)</span>
<span class="co">#&gt;     netdis3     netdis4 </span>
<span class="co">#&gt; 0.000000000 0.001268919</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<ul>
<li><p>W. Ali, T. Rito, G. Reinert, F. Sun, and C. M. Deane. Alignment-free protein interaction network comparison. Bioinformatics, 30:i430–i437, 2014.</p></li>
<li><p>L. Ospina-Forero, C. M. Deane, and G. Reinert. Assessment of model fit via network comparison methods based on subgraph counts. Journal of Complex Networks, page cny017, August 2018.</p></li>
<li><p>A. E. Wegner, L. Ospina-Forero, R. E. Gaunt, C. M. Deane, and G. Reinert. Identifying networks with common organizational principles. Journal of Complex networks, 2017.</p></li>
<li><p>F. Picard, J.-J. Daudin, M. Koskas, S. Schbath, and S. Robin. Assessing the exceptionality of network motifs. Journal of Computational Biology, 15(1):1–20, 2008.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin O’Reilly.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
