<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simple and quick (default) usage 2: pre-computed properties • netdist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simple and quick (default) usage 2: pre-computed properties">
<meta property="og:description" content="netdist">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">netdist</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ManyToMany.html">Simple and quick (default) usage 3: many to many comparisons</a>
    </li>
    <li>
      <a href="../articles/NetEmdTimeOrdering.html">NetEmd: World trade</a>
    </li>
    <li>
      <a href="../articles/NetdisGPStepByStep.html">Netdis-GP: step by step</a>
    </li>
    <li>
      <a href="../articles/NetdisStepByStep.html">Netdis: step by step</a>
    </li>
    <li>
      <a href="../articles/NewNetdisCustomisations.html">Netdis: customizations</a>
    </li>
    <li>
      <a href="../articles/PreComputedProps.html">Simple and quick (default) usage 2: pre-computed properties</a>
    </li>
    <li>
      <a href="../articles/V-Menu.html">Netdis Vignette's Menu</a>
    </li>
    <li>
      <a href="../articles/default_pairwise_usage.html">Simple and quick (default) usage 1: pairwise comparisons</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_dis.html">Netdis: Protein interaction networks</a>
    </li>
    <li>
      <a href="../articles/dendrogram_example_net_emd.html">Dendrogram example for NetEMD</a>
    </li>
    <li>
      <a href="../articles/netdis_customisations.html">Usage of netdis with binning and expected counts customisations.</a>
    </li>
    <li>
      <a href="../articles/netdis_pairwise_comparisons.html">Usage of netdis interfaces for different pairwise comparison options.</a>
    </li>
    <li>
      <a href="../articles/quickstart_netdis_2graphs.html">Quick start guide for Netdis - 2 graphs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simple and quick (default) usage 2: pre-computed properties</h1>
            
            <h4 data-toc-skip class="date">23/06/2020</h4>
      
      
      <div class="hidden name"><code>PreComputedProps.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In many situations there may be a need to pursue the network comparisons from pre-computed counts/features. Working from pre-computed features decreases the computational time to obtain results, particularly when the same networks are involved in multiple and subsequent comparisons, or when the same properties are used in other procedures. Similarly, advance users, could consider using other sets of pre-computed features, particularly for NetEmd.</p>
<p>This vignette follows similar examples as <a href="default_pairwise_usage.html">“Simple and quick (default) usage 1: pairwise comparisons”</a>. However, this vignette shows the steps required to obtain counts/features and then how to use them in the network comparison methods.</p>
<p>For other vignettes in this package see the <a href="V-Menu.html">“Menu”</a>.</p>
</div>
<div class="section level2">
<h2 id="load-required-packageslibraries">Load required packages/libraries<a class="anchor" aria-label="anchor" href="#load-required-packageslibraries"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load packages/libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"netdist"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://igraph.org" class="external-link">"igraph"</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compare-two-networks-via-netemd">Compare two networks via NetEmd<a class="anchor" aria-label="anchor" href="#compare-two-networks-via-netemd"></a>
</h2>
<p>The following sections show how to extract the subgraph counts, or other relevant features prior to running NetEmd or Netdis. Then it shows how to call NetEmd or Netdis using the pre-computed features.</p>
<div class="section level3">
<h3 id="comparing-two-graphs-with-netemd-via-subgraph-counts-">Comparing two graphs with NetEmd via subgraph counts.<a class="anchor" aria-label="anchor" href="#comparing-two-graphs-with-netemd-via-subgraph-counts-"></a>
</h3>
<p>NetEmd and Netdis use subgraph counts, however, NetEmd takes counts directly from the observed graph. Netdis takes counts from ego-networks of the observed graph. The following example shows how the counts are extracted for NetEmd via <code>count_orbits_per_node</code>. The user can call NetEmd by directly providing the matrix of counts, or by providing a processed version of these counts via <code>graph_features_to_histograms</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create lattice networks</span>
<span class="va">gLat_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span> 
<span class="va">gLat_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">44</span>,<span class="fl">44</span><span class="op">)</span><span class="op">)</span>  

<span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gLat_1</span>,vertex.size<span class="op">=</span><span class="fl">4</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gLat_2</span>,vertex.size<span class="op">=</span><span class="fl">4</span>,vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="PreComputedProps_files/figure-html/netemd-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The simple computation of NetEmd without pre-computed features:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/netemd_one_to_one.html">netemd_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span><span class="va">gLat_1</span>,graph_2<span class="op">=</span><span class="va">gLat_2</span>,feature_type<span class="op">=</span><span class="st">"orbit"</span>,smoothing_window_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.08389459</span></code></pre></div>
<div class="section level4">
<h4 id="providing-a-matrix-of-network-features">Providing a matrix of network features<a class="anchor" aria-label="anchor" href="#providing-a-matrix-of-network-features"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_1</span><span class="op">=</span> <span class="fu"><a href="../reference/count_orbits_per_node.html">count_orbits_per_node</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gLat_1</span>,max_graphlet_size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="va">counts_2</span><span class="op">=</span> <span class="fu"><a href="../reference/count_orbits_per_node.html">count_orbits_per_node</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gLat_2</span>,max_graphlet_size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">counts_1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;      O0 O1 O2 O3</span>
<span class="co">#&gt; [1,]  2  4  1  0</span>
<span class="co">#&gt; [2,]  3  6  3  0</span>
<span class="co">#&gt; [3,]  3  7  3  0</span>
<span class="co">#&gt; [4,]  3  7  3  0</span>
<span class="co">#&gt; [5,]  3  7  3  0</span>
<span class="co">#&gt; [6,]  3  7  3  0</span>

<span class="fu"><a href="../reference/netemd_one_to_one.html">netemd_one_to_one</a></span><span class="op">(</span>dhists_1<span class="op">=</span><span class="va">counts_1</span>, dhists_2<span class="op">=</span><span class="va">counts_2</span>,smoothing_window_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.08389459</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="providing-the-network-features-as-lists-of-dhist-objects">Providing the network features as lists of dhist objects<a class="anchor" aria-label="anchor" href="#providing-the-network-features-as-lists-of-dhist-objects"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dhists_1</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_features_to_histograms.html">graph_features_to_histograms</a></span><span class="op">(</span>features_matrix <span class="op">=</span> <span class="va">counts_1</span><span class="op">)</span>
<span class="va">dhists_2</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_features_to_histograms.html">graph_features_to_histograms</a></span><span class="op">(</span>features_matrix <span class="op">=</span> <span class="va">counts_2</span><span class="op">)</span>

<span class="fu"><a href="../reference/netemd_one_to_one.html">netemd_one_to_one</a></span><span class="op">(</span>dhists_1<span class="op">=</span><span class="va">dhists_1</span>, dhists_2<span class="op">=</span><span class="va">dhists_2</span>,smoothing_window_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.08389459</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="comparing-two-graphs-with-netemd-via-their-laplacian-spectrum">Comparing two graphs with NetEmd via their Laplacian spectrum<a class="anchor" aria-label="anchor" href="#comparing-two-graphs-with-netemd-via-their-laplacian-spectrum"></a>
</h3>
<p>Computation of the Laplacian and Normalized Laplacian:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Networks</span>
<span class="va">gLat_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span> 
<span class="va">gLat_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">44</span>,<span class="fl">44</span><span class="op">)</span><span class="op">)</span>

<span class="co">#Laplacian</span>
<span class="va">Lapg_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gLat_1</span>,normalized <span class="op">=</span> <span class="cn">FALSE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">Lapg_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gLat_2</span>,normalized <span class="op">=</span> <span class="cn">FALSE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#Normalized Laplacian</span>
<span class="va">NLapg_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gLat_1</span>,normalized <span class="op">=</span> <span class="cn">TRUE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">NLapg_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/laplacian_matrix.html" class="external-link">laplacian_matrix</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">gLat_2</span>,normalized <span class="op">=</span> <span class="cn">TRUE</span>,sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>


<span class="co"># Providing a matrix of network features (e.g. Spectra). (This may take a couple of minutes).</span>
<span class="va">spec_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>L.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">Lapg_1</span><span class="op">)</span><span class="op">$</span><span class="va">values</span>, NL.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">NLapg_1</span><span class="op">)</span><span class="op">$</span><span class="va">values</span><span class="op">)</span> 
<span class="va">spec_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>L.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">Lapg_2</span><span class="op">)</span><span class="op">$</span><span class="va">values</span>, NL.Spectra<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">NLapg_2</span><span class="op">)</span><span class="op">$</span><span class="va">values</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spec_1</span><span class="op">)</span>
<span class="co">#&gt;      L.Spectra NL.Spectra</span>
<span class="co">#&gt; [1,]  7.950753   2.000000</span>
<span class="co">#&gt; [2,]  7.877490   1.993354</span>
<span class="co">#&gt; [3,]  7.877490   1.993354</span>
<span class="co">#&gt; [4,]  7.804226   1.986361</span>
<span class="co">#&gt; [5,]  7.757390   1.973633</span>
<span class="co">#&gt; [6,]  7.757390   1.973633</span></code></pre></div>
<p>Similarly to counts, all other features can be given as a matrix or as dhist objects:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/netemd_one_to_one.html">netemd_one_to_one</a></span><span class="op">(</span>dhists_1 <span class="op">=</span> <span class="va">spec_1</span>,dhists_2 <span class="op">=</span> <span class="va">spec_2</span>, smoothing_window_width <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.01280812</span>


<span class="co"># Providing pre-computed dhist objects from network features</span>
<span class="va">dhists_1</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_features_to_histograms.html">graph_features_to_histograms</a></span><span class="op">(</span><span class="va">spec_1</span><span class="op">)</span>
<span class="va">dhists_2</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_features_to_histograms.html">graph_features_to_histograms</a></span><span class="op">(</span><span class="va">spec_2</span><span class="op">)</span>

<span class="fu"><a href="../reference/netemd_one_to_one.html">netemd_one_to_one</a></span><span class="op">(</span>dhists_1<span class="op">=</span><span class="va">dhists_1</span>, dhists_2<span class="op">=</span><span class="va">dhists_2</span>, smoothing_window_width <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.01280812</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="compare-two-networks-via-netdis-and-its-variants">Compare two networks via Netdis and its variants<a class="anchor" aria-label="anchor" href="#compare-two-networks-via-netdis-and-its-variants"></a>
</h2>
<p>Netdis uses counts from the resulting ego-networks of each of the nodes in a graph, but it also considers the mean or expectation of these ego-network counts (<span class="math inline">\(E_w\)</span>). The following example shows how the user can compute and provide the properties required from each graph as well as any relevant property required to calculate <span class="math inline">\(E_w\)</span>. The main properties can be obtained by means of <code>count_graphlets_ego</code>.</p>
<div class="section level3">
<h3 id="using-netdis-with-a-gold-standard-or-reference-graph-to-obtain-e_w">Using Netdis with a gold-standard or reference graph to obtain <span class="math inline">\(E_w\)</span><a class="anchor" aria-label="anchor" href="#using-netdis-with-a-gold-standard-or-reference-graph-to-obtain-e_w"></a>
</h3>
<p>The selection of a gold-standard graph as a substitute for <span class="math inline">\(E_w\)</span> could be done when such graph is known to be a good proxy for <span class="math inline">\(E_w\)</span>, or alternatively as a good reference point for the comparison. This option will focus on detecting discrepancies between the networks relative to the ego-network structure of the reference network / gold-standard graph and which is summarized in <span class="math inline">\(E_w\)</span>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set source directory for virus protein-protein interaction edge files stored in the netdist package.</span>
<span class="va">source_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"VRPINS"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"netdist"</span><span class="op">)</span>

<span class="co"># Load query graphs as igraph objects</span>
<span class="va">graph_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"EBV.txt"</span><span class="op">)</span>,format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>
<span class="va">graph_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"ECL.txt"</span><span class="op">)</span>,format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span></code></pre></div>
<p>For illustration purposes consider the lattice networks as possible gold-standard networks:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Lattice graphs to be used as gold-standard as a reference point comparison</span>
<span class="va">goldstd_1</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span> <span class="co">#Graph with 8^2 nodes</span>
<span class="va">goldstd_2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/make_lattice.html" class="external-link">graph.lattice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">44</span>,<span class="fl">44</span><span class="op">)</span><span class="op">)</span> <span class="co">#Graph with 44^2 nodes</span></code></pre></div>
<p>Now obtain the subgraph counts for all networks.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Providing pre-calculated subgraph counts.</span>
<span class="va">props_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_1</span><span class="op">)</span>
<span class="va">props_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_2</span><span class="op">)</span>
<span class="va">props_goldstd_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">goldstd_1</span><span class="op">)</span>
<span class="va">props_goldstd_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">goldstd_2</span><span class="op">)</span></code></pre></div>
<p>Compute Netdis using the pre-computed counts and any of the example gold-standard networks.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graphlet_counts_1<span class="op">=</span> <span class="va">props_1</span>,graphlet_counts_2<span class="op">=</span> <span class="va">props_2</span>, graphlet_counts_ref <span class="op">=</span> <span class="va">props_goldstd_1</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.1405770 0.2507667</span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graphlet_counts_1<span class="op">=</span> <span class="va">props_1</span>,graphlet_counts_2<span class="op">=</span> <span class="va">props_2</span>, graphlet_counts_ref <span class="op">=</span> <span class="va">props_goldstd_2</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.1401654 0.2505384</span></code></pre></div>
<p>Comparison to the result of Netdis without pre-computed counts.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Netdis using the goldstd_1 graph as gold-standard reference point</span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="va">goldstd_1</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.1405770 0.2507667</span>

<span class="co"># Netdis using the goldstd_2 graph as gold-standard reference point</span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="va">goldstd_2</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.1401654 0.2505384</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="netdis-gp-using-a-geometric-poisson-approximation">Netdis-GP: Using a Geometric-Poisson approximation<a class="anchor" aria-label="anchor" href="#netdis-gp-using-a-geometric-poisson-approximation"></a>
</h3>
<p>This Netdis variant focuses on detecting more meso-level discrepancies between the ego-network structures.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set source directory for Virus protein-protein interaction edge files stored in the netdist package.</span>
<span class="va">source_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"VRPINS"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"netdist"</span><span class="op">)</span>

<span class="co"># Load query graphs as igraph objects</span>
<span class="va">graph_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"EBV.txt"</span><span class="op">)</span>,format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>
<span class="va">graph_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"ECL.txt"</span><span class="op">)</span>,format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>

<span class="co"># Netdis using the Geometric-Poisson approximation as a way to obtain background expectations. </span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.8822527 0.1892716</span>

<span class="co"># Providing pre-calculated subgraph counts.</span>
<span class="va">props_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_1</span><span class="op">)</span>
<span class="va">props_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_2</span><span class="op">)</span>

<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graphlet_counts_1<span class="op">=</span> <span class="va">props_1</span>,graphlet_counts_2<span class="op">=</span> <span class="va">props_2</span>, ref_graph <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="co">#&gt;   netdis3   netdis4 </span>
<span class="co">#&gt; 0.8822527 0.1892716</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-netdis-with-no-background-expectation-e_w0">Using Netdis with no background expectation (<span class="math inline">\(E_w=0\)</span>)<a class="anchor" aria-label="anchor" href="#using-netdis-with-no-background-expectation-e_w0"></a>
</h3>
<p>Comparing the networks via their observed ego counts without centring them, (equivalent to using expectation equal to zero). This variant focuses on detecting small discrepancies between the networks.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set source directory for Virus protein-protein interaction edge files stored in the netdist package.</span>
<span class="va">source_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"VRPINS"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"netdist"</span><span class="op">)</span>

<span class="co"># Load query graphs as igraph objects</span>
<span class="va">graph_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"EBV.txt"</span><span class="op">)</span>,format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>
<span class="va">graph_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_simple_graph.html">read_simple_graph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">source_dir</span>, <span class="st">"ECL.txt"</span><span class="op">)</span>,format <span class="op">=</span> <span class="st">"ncol"</span><span class="op">)</span>

<span class="co">#Netdis using no expectations (or equivalently, expectation equal to zero).</span>
<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graph_1<span class="op">=</span> <span class="va">graph_1</span>, graph_2<span class="op">=</span> <span class="va">graph_2</span>,  ref_graph <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt;    netdis3    netdis4 </span>
<span class="co">#&gt; 0.00761545 0.02106628</span>

<span class="co"># Providing pre-calculated subgraph counts.</span>
<span class="va">props_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_1</span><span class="op">)</span>
<span class="va">props_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_graphlets_ego.html">count_graphlets_ego</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph_2</span><span class="op">)</span>

<span class="fu"><a href="../reference/netdis_one_to_one.html">netdis_one_to_one</a></span><span class="op">(</span>graphlet_counts_1<span class="op">=</span> <span class="va">props_1</span>,graphlet_counts_2<span class="op">=</span> <span class="va">props_2</span>, ref_graph <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>  
<span class="co">#&gt;    netdis3    netdis4 </span>
<span class="co">#&gt; 0.00761545 0.02106628</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<ul>
<li><p>W. Ali, T. Rito, G. Reinert, F. Sun, and C. M. Deane. Alignment-free protein interaction network comparison. Bioinformatics, 30:i430–i437, 2014.</p></li>
<li><p>L. Ospina-Forero, C. M. Deane, and G. Reinert. Assessment of model fit via network comparison methods based on subgraph counts. Journal of Complex Networks, page cny017, August 2018.</p></li>
<li><p>A. E. Wegner, L. Ospina-Forero, R. E. Gaunt, C. M. Deane, and G. Reinert. Identifying networks with common organizational principles. Journal of Complex networks, 2017.</p></li>
<li><p>F. Picard, J.-J. Daudin, M. Koskas, S. Schbath, and S. Robin. Assessing the exceptionality of network motifs. Journal of Computational Biology, 15(1):1–20, 2008.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin O’Reilly.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
